// @syncfusion/ej2-pdf-export/src/implementation/general/pdf-cache-collection.js
define(["require", "exports", "./../collections/dictionary"], function (require, exports, dictionary_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfCacheCollection = (function () {
        function PdfCacheCollection() {
            this.referenceObjects = [];
            this.pdfFontCollection = new dictionary_1.Dictionary();
        }
        PdfCacheCollection.prototype.search = function (obj) {
            var result = null;
            var group = this.createNewGroup();
            group.push(obj);
            return result;
        };
        PdfCacheCollection.prototype.createNewGroup = function () {
            var group = [];
            this.referenceObjects.push(group);
            return group;
        };
        PdfCacheCollection.prototype.destroy = function () {
            this.pdfFontCollection = undefined;
            this.referenceObjects = undefined;
        };
        return PdfCacheCollection;
    }());
    exports.PdfCacheCollection = PdfCacheCollection;
});
