// @syncfusion/ej2-grids/src/grid/renderer/default-edit-cell.js
define(["require", "exports", "@syncfusion/ej2-base", "@syncfusion/ej2-inputs", "../base/util"], function (require, exports, ej2_base_1, ej2_inputs_1, util_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DefaultEditCell = (function () {
        function DefaultEditCell(parent) {
            this.parent = parent;
        }
        DefaultEditCell.prototype.create = function (args) {
            var col = args.column;
            var input = ej2_base_1.createElement('input', {
                className: 'e-field e-input e-defaultcell', attrs: {
                    type: 'text', value: !ej2_base_1.isNullOrUndefined(args.value) ? args.value : '', 'e-mappinguid': col.uid,
                    id: this.parent.element.id + col.field, name: col.field, style: 'text-align:' + col.textAlign,
                }
            });
            return input;
        };
        DefaultEditCell.prototype.read = function (element) {
            return element.value;
        };
        DefaultEditCell.prototype.write = function (args) {
            var col = args.column;
            var isInline = this.parent.editSettings.mode !== 'Dialog';
            ej2_inputs_1.Input.createInput({
                element: args.element, floatLabelType: this.parent.editSettings.mode !== 'Dialog' ? 'Never' : 'Always',
                properties: {
                    enableRtl: this.parent.enableRtl, enabled: util_1.isEditable(args.column, args.requestType, args.element),
                    placeholder: isInline ? '' : args.column.headerText
                }
            });
        };
        return DefaultEditCell;
    }());
    exports.DefaultEditCell = DefaultEditCell;
});
