// @syncfusion/ej2-grids/src/grid/models/aggregate.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
define(["require", "exports", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "../base/enum", "@syncfusion/ej2-base", "../services/value-formatter"], function (require, exports, ej2_base_1, ej2_base_2, enum_1, ej2_base_3, value_formatter_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AggregateColumn = (function (_super) {
        __extends(AggregateColumn, _super);
        function AggregateColumn() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.templateFn = {};
            return _this;
        }
        AggregateColumn.prototype.setFormatter = function () {
            var valueFormatter = new value_formatter_1.ValueFormatter();
            if (this.format && (this.format.skeleton || this.format.format)) {
                this.formatFn = valueFormatter.getFormatFunction(this.format);
            }
        };
        AggregateColumn.prototype.getFormatter = function () {
            return this.formatFn;
        };
        AggregateColumn.prototype.setTemplate = function (helper) {
            if (helper === void 0) { helper = {}; }
            if (this.footerTemplate !== undefined) {
                this.templateFn[ej2_base_2.getEnumValue(enum_1.CellType, enum_1.CellType.Summary)] = { fn: ej2_base_1.compile(this.footerTemplate, helper),
                    property: 'footerTemplate' };
            }
            if (this.groupFooterTemplate !== undefined) {
                this.templateFn[ej2_base_2.getEnumValue(enum_1.CellType, enum_1.CellType.GroupSummary)] = { fn: ej2_base_1.compile(this.groupFooterTemplate, helper),
                    property: 'groupFooterTemplate' };
            }
            if (this.groupCaptionTemplate !== undefined) {
                this.templateFn[ej2_base_2.getEnumValue(enum_1.CellType, enum_1.CellType.CaptionSummary)] = { fn: ej2_base_1.compile(this.groupCaptionTemplate, helper),
                    property: 'groupCaptionTemplate' };
            }
        };
        AggregateColumn.prototype.getTemplate = function (type) {
            return this.templateFn[ej2_base_2.getEnumValue(enum_1.CellType, type)];
        };
        AggregateColumn.prototype.setPropertiesSilent = function (prop) {
            this.setProperties(prop, true);
        };
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "type", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "field", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "columnName", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "format", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "footerTemplate", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "groupFooterTemplate", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "groupCaptionTemplate", void 0);
        __decorate([
            ej2_base_3.Property()
        ], AggregateColumn.prototype, "customAggregate", void 0);
        return AggregateColumn;
    }(ej2_base_3.ChildProperty));
    exports.AggregateColumn = AggregateColumn;
    var AggregateRow = (function (_super) {
        __extends(AggregateRow, _super);
        function AggregateRow() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        __decorate([
            ej2_base_3.Collection([], AggregateColumn)
        ], AggregateRow.prototype, "columns", void 0);
        return AggregateRow;
    }(ej2_base_3.ChildProperty));
    exports.AggregateRow = AggregateRow;
});
