// @syncfusion/ej2-ng-inputs/src/maskedtextbox/maskedtextbox.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, Renderer2, ChangeDetectionStrategy, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ComponentBase, applyMixins, FormBase, setValue } from '@syncfusion/ej2-ng-base';
import { MaskedTextBox } from '@syncfusion/ej2-inputs';
export var inputs = ['cssClass', 'customCharacters', 'enablePersistence', 'enableRtl', 'enabled', 'floatLabelType', 'locale', 'mask', 'placeholder', 'promptChar', 'showClearButton', 'value', 'width'];
export var outputs = ['change', 'created', 'destroyed', 'focus', 'valueChange'];
export var twoWays = ['value'];
/**
 * Represents the EJ2 Angular MaskedTextbox Component.
 * ```html
 * <ej-maskedtextbox [value]='value'></ej-maskedtextbox>
 * ```
 */
var MaskedTextBoxComponent = /** @class */ (function (_super) {
    __extends(MaskedTextBoxComponent, _super);
    function MaskedTextBoxComponent(ngEle, srenderer, viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.srenderer = srenderer;
        _this.viewContainerRef = viewContainerRef;
        _this.skipFromEvent = true;
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 3; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    MaskedTextBoxComponent.prototype.registerOnChange = function (registerFunction) {
    };
    MaskedTextBoxComponent.prototype.registerOnTouched = function (registerFunction) {
    };
    MaskedTextBoxComponent.prototype.writeValue = function (value) {
    };
    MaskedTextBoxComponent.prototype.setDisabledState = function (disabled) {
    };
    MaskedTextBoxComponent.prototype.ngOnInit = function () {
    };
    MaskedTextBoxComponent.prototype.ngAfterViewInit = function () {
    };
    MaskedTextBoxComponent.prototype.ngOnDestroy = function () {
    };
    MaskedTextBoxComponent.prototype.ngAfterContentChecked = function () {
    };
    MaskedTextBoxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ejs-maskedtextbox',
                    inputs: inputs,
                    outputs: outputs,
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return MaskedTextBoxComponent; }),
                            multi: true
                        }
                    ],
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    MaskedTextBoxComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: ViewContainerRef, },
    ]; };
    return MaskedTextBoxComponent;
}(MaskedTextBox));
export { MaskedTextBoxComponent };
applyMixins(MaskedTextBoxComponent, [ComponentBase, FormBase]);
