// @syncfusion/ej2-dropdowns/src/multi-select/multi-select.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
define(["require", "exports", "../drop-down-base/drop-down-base", "@syncfusion/ej2-popups", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "../common/incremental-search", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-data", "./float-label"], function (require, exports, drop_down_base_1, ej2_popups_1, ej2_base_1, ej2_base_2, ej2_base_3, ej2_base_4, incremental_search_1, ej2_base_5, ej2_base_6, ej2_data_1, float_label_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FOCUS = 'e-input-focus';
    var DISABLED = 'e-disabled';
    var OVER_ALL_WRAPPER = 'e-multiselect e-input-group';
    var ELEMENT_WRAPPER = 'e-multi-select-wrapper';
    var ELEMENT_MOBILE_WRAPPER = 'e-mob-wrapper';
    var HIDE_LIST = 'e-hide-listitem';
    var DELIMITER_VIEW = 'e-delim-view';
    var CHIP_WRAPPER = 'e-chips-collection';
    var CHIP = 'e-chips';
    var CHIP_CONTENT = 'e-chipcontent';
    var CHIP_CLOSE = 'e-chips-close';
    var CHIP_SELECTED = 'e-chip-selected';
    var SEARCHBOX_WRAPPER = 'e-searcher';
    var DELIMITER_VIEW_WRAPPER = 'e-delimiter';
    var ZERO_SIZE = 'e-zero-size';
    var REMAIN_WRAPPER = 'e-remain';
    var CLOSEICON_CLASS = 'e-chips-close e-close-hooker';
    var DELIMITER_WRAPPER = 'e-delim-values';
    var POPUP_WRAPPER = 'e-ddl e-popup e-multi-select-list-wrapper';
    var INPUT_ELEMENT = 'e-dropdownbase';
    var RTL_CLASS = 'e-rtl';
    var CLOSE_ICON_HIDE = 'e-close-icon-hide';
    var MOBILE_CHIP = 'e-mob-chip';
    var FOOTER = 'e-ddl-footer';
    var HEADER = 'e-ddl-header';
    var DISABLE_ICON = 'e-ddl-disable-icon';
    var SPINNER_CLASS = 'e-ms-spinner-icon';
    var HIDDEN_ELEMENT = 'e-multi-hidden';
    var destroy = 'destroy';
    var dropdownIcon = 'e-input-group-icon e-ddl-icon';
    var iconAnimation = 'e-icon-anim';
    var MultiSelect = (function (_super) {
        __extends(MultiSelect, _super);
        function MultiSelect(option, element) {
            var _this = _super.call(this, option, element) || this;
            _this.mobFilter = true;
            _this.isFiltered = false;
            _this.focused = true;
            _this.isValidKey = false;
            _this.scrollFocusStatus = false;
            _this.keyDownStatus = false;
            return _this;
        }
        ;
        MultiSelect.prototype.enableRTL = function (state) {
            if (state) {
                this.overAllWrapper.classList.add(RTL_CLASS);
            }
            else {
                this.overAllWrapper.classList.remove(RTL_CLASS);
            }
            if (this.popupObj) {
                this.popupObj.enableRtl = state;
                this.popupObj.dataBind();
            }
        };
        MultiSelect.prototype.requiredModules = function () {
            var modules = [];
            if (this.mode === 'CheckBox') {
                this.allowCustomValue = false;
                this.hideSelectedItem = false;
                this.closePopupOnSelect = false;
                this.allowFiltering = true;
                modules.push({
                    member: 'CheckBoxSelection',
                    args: [this]
                });
            }
            return modules;
        };
        MultiSelect.prototype.updateHTMLAttribute = function () {
            if (Object.keys(this.htmlAttributes).length) {
                for (var _i = 0, _a = Object.keys(this.htmlAttributes); _i < _a.length; _i++) {
                    var htmlAttr = _a[_i];
                    switch (htmlAttr) {
                        case 'class':
                            this.overAllWrapper.classList.add(this.htmlAttributes[htmlAttr]);
                            this.popupWrapper.classList.add(this.htmlAttributes[htmlAttr]);
                            break;
                        case 'disabled':
                            this.enable(false);
                            break;
                        case 'placeholder':
                            this.inputElement.setAttribute(htmlAttr, this.htmlAttributes[htmlAttr]);
                            break;
                        default:
                            var defaultAttr = ['title', 'id'];
                            var validateAttr = ['name', 'required'];
                            if (validateAttr.indexOf(htmlAttr) > -1) {
                                this.hiddenElement.setAttribute(htmlAttr, this.htmlAttributes[htmlAttr]);
                            }
                            else if (defaultAttr.indexOf(htmlAttr) > -1) {
                                this.element.setAttribute(htmlAttr, this.htmlAttributes[htmlAttr]);
                            }
                            else {
                                this.overAllWrapper.setAttribute(htmlAttr, this.htmlAttributes[htmlAttr]);
                            }
                            break;
                    }
                }
            }
        };
        MultiSelect.prototype.updateReadonly = function (state) {
            if (state || this.mode === 'CheckBox') {
                this.inputElement.setAttribute('readonly', 'true');
            }
            else {
                this.inputElement.removeAttribute('readonly');
            }
        };
        MultiSelect.prototype.updateClearButton = function (state) {
            if (state) {
                if (this.overAllClear.parentNode) {
                    this.overAllClear.style.display = '';
                }
                else {
                    this.componentWrapper.appendChild(this.overAllClear);
                }
                this.componentWrapper.classList.remove(CLOSE_ICON_HIDE);
            }
            else {
                this.overAllClear.style.display = 'none';
                this.componentWrapper.classList.add(CLOSE_ICON_HIDE);
            }
        };
        MultiSelect.prototype.updateCssClass = function () {
            if (this.cssClass) {
                this.popupWrapper.classList.add(this.cssClass);
                this.overAllWrapper.classList.add(this.cssClass);
            }
        };
        MultiSelect.prototype.onPopupShown = function () {
            if (ej2_base_4.Browser.isDevice && (this.mode === 'CheckBox' && this.allowFiltering)) {
                var proxy_1 = this;
                window.onpopstate = function () {
                    proxy_1.hidePopup();
                };
                history.pushState({}, '');
            }
            var eventArgs = { popup: this.popupObj, cancel: false };
            this.trigger('open', eventArgs);
            if (eventArgs.cancel) {
                return;
            }
            var animModel = { name: 'FadeIn', duration: 100 };
            this.focusAtFirstListItem();
            document.body.appendChild(this.popupObj.element);
            this.refreshPopup();
            if (this.mode === 'CheckBox') {
                ej2_base_5.addClass([this.overAllWrapper], [iconAnimation]);
            }
            this.popupObj.show(animModel);
            ej2_base_1.attributes(this.inputElement, { 'aria-expanded': 'true' });
            if (!this.isFirstClick) {
                var ulElement = this.list.querySelector('ul');
                if (ulElement) {
                    this.mainList = ulElement.cloneNode ? ulElement.cloneNode(true) : ulElement;
                }
                this.isFirstClick = true;
            }
            this.refreshListItems(null);
            this.notify('reOrder', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', e: this });
        };
        MultiSelect.prototype.focusAtFirstListItem = function () {
            if (this.ulElement && this.ulElement.querySelector('li.'
                + drop_down_base_1.dropDownBaseClasses.li)) {
                var element = void 0;
                if (this.mode === 'CheckBox') {
                    element = this.list.querySelector('li.'
                        + drop_down_base_1.dropDownBaseClasses.li + ':not(.'
                        + HIDE_LIST + ')');
                }
                else {
                    element = this.ulElement.querySelector('li.'
                        + drop_down_base_1.dropDownBaseClasses.li + ':not(.'
                        + HIDE_LIST + ')');
                }
                if (element !== null) {
                    this.removeFocus();
                    this.addListFocus(element);
                }
            }
        };
        MultiSelect.prototype.focusAtLastListItem = function (data) {
            var activeElement;
            if (data) {
                activeElement = incremental_search_1.Search(data, this.liCollections, 'StartsWith', this.ignoreCase);
            }
            else {
                if (this.value && this.value.length) {
                    incremental_search_1.Search(this.value[this.value.length - 1], this.liCollections, 'StartsWith', this.ignoreCase);
                }
                else {
                    activeElement = null;
                }
            }
            if (activeElement && activeElement.item !== null) {
                this.addListFocus(activeElement.item);
                this.scrollBottom(activeElement.item, activeElement.index);
            }
        };
        MultiSelect.prototype.getAriaAttributes = function () {
            var ariaAttributes = {
                'aria-disabled': 'false',
                'aria-owns': this.element.id + '_options',
                'role': 'listbox',
                'aria-multiselectable': 'true',
                'aria-activedescendant': 'null',
                'aria-haspopup': 'true',
                'aria-expanded': 'false'
            };
            return ariaAttributes;
        };
        MultiSelect.prototype.updateListARIA = function () {
            ej2_base_1.attributes(this.ulElement, { 'id': this.element.id + '_options', 'role': 'listbox', 'aria-hidden': 'false' });
            ej2_base_1.attributes(this.inputElement, this.getAriaAttributes());
            var li;
            li = this.list.querySelectorAll('li.' + drop_down_base_1.dropDownBaseClasses.li);
            var temp = li.length;
            if (li && li.length) {
                while (temp > 0) {
                    if (li[temp - 1].getAttribute('aria-selected') !== 'true') {
                        li[temp - 1].setAttribute('aria-selected', 'false');
                    }
                    temp--;
                }
            }
        };
        MultiSelect.prototype.removelastSelection = function (e) {
            var elements;
            elements = this.chipCollectionWrapper.querySelectorAll('span.' + CHIP);
            var value = elements[elements.length - 1].getAttribute('data-value');
            if (!ej2_base_6.isNullOrUndefined(this.value)) {
                this.tempValues = this.value.slice();
            }
            this.removeValue(value, e);
            this.removeChipSelection();
            this.updateDelimeter(this.delimiterChar);
            this.makeTextBoxEmpty();
            if (this.allowFiltering && this.mainList && this.listData) {
                var list = this.mainList.cloneNode ? this.mainList.cloneNode(true) : this.mainList;
                this.onActionComplete(list, this.mainData);
                this.refreshSelection();
            }
            this.focusAtLastListItem(value);
        };
        MultiSelect.prototype.onActionFailure = function (e) {
            _super.prototype.onActionFailure.call(this, e);
            this.renderPopup();
            this.onPopupShown();
        };
        MultiSelect.prototype.targetElement = function () {
            this.targetInputElement = this.inputElement;
            if (this.mode === 'CheckBox') {
                this.notify('targetElement', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox' });
            }
            return this.targetInputElement.value;
        };
        MultiSelect.prototype.onActionComplete = function (ulElement, list, e, isUpdated) {
            _super.prototype.onActionComplete.call(this, ulElement, list, e);
            if (this.mode === 'CheckBox' && this.showSelectAll) {
                this.notify('selectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox' });
            }
            if (!this.mainList && !this.mainData) {
                this.mainList = ulElement.cloneNode ? ulElement.cloneNode(true) : ulElement;
                this.mainData = list;
                this.mainListCollection = this.liCollections;
            }
            else if (!ej2_base_6.isNullOrUndefined(this.mainData) && this.mainData.length === 0) {
                this.mainData = list;
            }
            if ((this.remoteCustomValue || list.length <= 0) && this.allowCustomValue && this.inputFocus && this.allowFiltering) {
                this.checkForCustomValue(this.tempQuery, this.fields);
                return;
            }
            if (this.value && this.value.length) {
                this.refreshSelection();
            }
            this.updateListARIA();
            this.unwireListEvents();
            this.wireListEvents();
            if (!ej2_base_6.isNullOrUndefined(this.setInitialValue)) {
                this.setInitialValue();
            }
            if (!ej2_base_6.isNullOrUndefined(this.selectAllAction)) {
                this.selectAllAction();
            }
            if (this.setDynValue) {
                this.initialValueUpdate();
                this.initialUpdate();
                this.refreshPlaceHolder();
                this.updateValueState(null, this.value, null);
            }
            this.renderPopup();
            this.refreshPopup();
            if (this.beforePopupOpen) {
                this.beforePopupOpen = false;
                this.onPopupShown();
            }
        };
        MultiSelect.prototype.refreshSelection = function () {
            var value;
            var element;
            var className = this.hideSelectedItem ?
                HIDE_LIST :
                drop_down_base_1.dropDownBaseClasses.selected;
            if (!ej2_base_6.isNullOrUndefined(this.value)) {
                for (var index = 0; this.value[index]; index++) {
                    value = this.value[index];
                    element = this.ulElement.querySelector('li[data-value="' + value + '"]');
                    if (element) {
                        ej2_base_5.addClass([element], className);
                        if (this.hideSelectedItem && element.previousSibling
                            && element.previousElementSibling.classList.contains(drop_down_base_1.dropDownBaseClasses.group)
                            && (!element.nextElementSibling ||
                                element.nextElementSibling.classList.contains(drop_down_base_1.dropDownBaseClasses.group))) {
                            ej2_base_5.addClass([element.previousElementSibling], className);
                        }
                        if (this.hideSelectedItem && element.classList.contains(drop_down_base_1.dropDownBaseClasses.focus)) {
                            ej2_base_5.removeClass([element], drop_down_base_1.dropDownBaseClasses.focus);
                            var listEle = element.parentElement.querySelectorAll('.' +
                                drop_down_base_1.dropDownBaseClasses.li + ':not(.' + HIDE_LIST + ')');
                            if (listEle.length > 0) {
                                ej2_base_5.addClass([listEle[0]], drop_down_base_1.dropDownBaseClasses.focus);
                            }
                            else {
                                this.l10nUpdate();
                                ej2_base_5.addClass([this.list], drop_down_base_1.dropDownBaseClasses.noData);
                            }
                        }
                        element.setAttribute('aria-selected', 'true');
                        if (this.mode === 'CheckBox' && element.classList.contains('e-active')) {
                            var ariaValue = element.firstElementChild.getAttribute('aria-checked');
                            if (ej2_base_6.isNullOrUndefined(ariaValue) || ariaValue === 'false') {
                                var args = {
                                    module: 'CheckBoxSelection',
                                    enable: this.mode === 'CheckBox',
                                    li: element,
                                    e: null
                                };
                                this.notify('updatelist', args);
                            }
                        }
                    }
                }
            }
            this.checkSelectAll();
            this.checkMaxSelection();
        };
        MultiSelect.prototype.checkSelectAll = function () {
            var searchCount = this.list.querySelectorAll('li.' + drop_down_base_1.dropDownBaseClasses.li).length;
            var searchActiveCount = this.list.querySelectorAll('li.' + drop_down_base_1.dropDownBaseClasses.selected).length;
            if ((searchCount === searchActiveCount) && (this.mode === 'CheckBox' && this.showSelectAll)) {
                this.notify('checkSelectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', value: 'check' });
            }
            if ((searchCount !== searchActiveCount) && (this.mode === 'CheckBox' && this.showSelectAll)) {
                this.notify('checkSelectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', value: 'uncheck' });
            }
        };
        MultiSelect.prototype.openClick = function (e) {
            if (!this.openOnClick) {
                if (this.targetElement() !== '') {
                    this.showPopup();
                }
                else {
                    this.hidePopup();
                }
            }
        };
        MultiSelect.prototype.KeyUp = function (e) {
            var _this = this;
            this.isValidKey = e.keyCode === 8 || this.isValidKey;
            if (this.isValidKey) {
                this.isValidKey = false;
                this.expandTextbox();
                switch (e.keyCode) {
                    default:
                        if (!this.isPopupOpen() && this.openOnClick) {
                            this.showPopup();
                        }
                        this.openClick(e);
                        if (this.checkTextLength() && !this.allowFiltering && (e.keyCode !== 8)) {
                            this.focusAtFirstListItem();
                        }
                        else {
                            var text = this.targetElement();
                            this.keyCode = e.keyCode;
                            if (this.allowFiltering) {
                                var eventArgs_1 = {
                                    preventDefaultAction: false,
                                    text: this.targetElement(),
                                    updateData: function (dataSource, query, fields) {
                                        if (eventArgs_1.cancel) {
                                            return;
                                        }
                                        _this.isFiltered = true;
                                        _this.remoteFilterAction = true;
                                        _this.dataUpdater(dataSource, query, fields);
                                    },
                                    event: e,
                                    cancel: false
                                };
                                this.trigger('filtering', eventArgs_1);
                                if (eventArgs_1.cancel) {
                                    return;
                                }
                                if (!this.isFiltered && !eventArgs_1.preventDefaultAction) {
                                    this.filterAction = true;
                                    this.dataUpdater(this.dataSource, null, this.fields);
                                }
                            }
                            else if (this.allowCustomValue) {
                                var query = new ej2_data_1.Query();
                                query = (text !== '') ? query.where(this.fields.text, 'startswith', text, true, this.ignoreAccent) : query;
                                this.dataUpdater(this.mainData, query, this.fields);
                                break;
                            }
                            else {
                                var liCollections = void 0;
                                liCollections = this.list.querySelectorAll('li.' + drop_down_base_1.dropDownBaseClasses.li + ':not(.e-hide-listitem)');
                                var activeElement = incremental_search_1.Search(this.targetElement(), liCollections, 'StartsWith', this.ignoreCase);
                                if (activeElement && activeElement.item !== null) {
                                    this.addListFocus(activeElement.item);
                                    this.list.scrollTop =
                                        activeElement.item.getBoundingClientRect().height * activeElement.index;
                                }
                                else if (this.targetElement() !== '') {
                                    this.removeFocus();
                                }
                                else {
                                    this.focusAtFirstListItem();
                                }
                            }
                        }
                }
            }
        };
        MultiSelect.prototype.getQuery = function (query) {
            var filterQuery = query ? query.clone() : this.query ? this.query.clone() : new ej2_data_1.Query();
            if (this.filterAction) {
                if (this.targetElement() !== null) {
                    var dataType = this.typeOfData(this.dataSource).typeof;
                    if (!(this.dataSource instanceof ej2_data_1.DataManager) && dataType === 'string' || dataType === 'number') {
                        filterQuery.where('', 'startswith', this.targetElement(), true, this.ignoreAccent);
                    }
                    else {
                        var fields = this.fields;
                        filterQuery.where(!ej2_base_6.isNullOrUndefined(fields.text) ? fields.text : '', 'startswith', this.targetElement(), true, this.ignoreAccent);
                    }
                }
                return filterQuery;
            }
            else {
                if (this.dataSource instanceof ej2_data_1.DataManager && !ej2_base_6.isNullOrUndefined(this.value) && !this.remoteFilterAction) {
                    var predicate = void 0;
                    var field = ej2_base_6.isNullOrUndefined(this.fields.value) ? this.fields.text : this.fields.value;
                    for (var i = 0; i < this.value.length; i++) {
                        if (i === 0) {
                            predicate = new ej2_data_1.Predicate(field, 'equal', this.value[i]);
                        }
                        else {
                            predicate = predicate.or(field, 'equal', this.value[i]);
                        }
                    }
                    return new ej2_data_1.Query().where(predicate);
                }
                else {
                    return query ? query : this.query ? this.query : new ej2_data_1.Query();
                }
            }
        };
        MultiSelect.prototype.dataUpdater = function (dataSource, query, fields) {
            if (this.targetElement().trim() === '') {
                var list = this.mainList.cloneNode ? this.mainList.cloneNode(true) : this.mainList;
                this.remoteCustomValue = false;
                this.onActionComplete(list, this.mainData);
                if (this.value && this.value.length) {
                    this.refreshSelection();
                }
                if (this.keyCode !== 8) {
                    this.focusAtFirstListItem();
                }
                this.notify('reOrder', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', e: this });
            }
            else {
                this.resetList(dataSource, fields, query);
                if (this.allowCustomValue) {
                    if (!(dataSource instanceof ej2_data_1.DataManager)) {
                        this.checkForCustomValue(query, fields);
                    }
                    else {
                        this.remoteCustomValue = true;
                        this.tempQuery = query;
                    }
                }
            }
            this.refreshPopup();
        };
        MultiSelect.prototype.checkForCustomValue = function (query, fields) {
            var dataChecks = !this.getValueByText(this.inputElement.value, this.ignoreCase);
            if (this.allowCustomValue && dataChecks) {
                var text = this.fields.text;
                var value = this.fields.value;
                var data = {};
                var customData = (!ej2_base_6.isNullOrUndefined(this.mainData) && this.mainData.length > 0) ?
                    this.mainData[0] : this.mainData;
                if (typeof (customData) !== 'string') {
                    var dataSource = Object.keys(customData);
                    for (var i = 0; i < dataSource.length; i++) {
                        data[dataSource[i]] = '';
                    }
                    data[value] = data[text] = this.inputElement.value;
                    var tempData = JSON.parse(JSON.stringify(this.listData));
                    tempData.splice(0, 0, data);
                    this.resetList(tempData, fields ? fields : this.fields, query);
                }
                else {
                    var tempData = [this.inputElement.value];
                    this.resetList(tempData, fields ? fields : this.fields);
                }
            }
            if (this.value && this.value.length) {
                this.refreshSelection();
            }
        };
        MultiSelect.prototype.getNgDirective = function () {
            return 'EJS-MULTISELECT';
        };
        MultiSelect.prototype.wrapperClick = function (e) {
            this.setDynValue = false;
            if (this.readonly || !this.enabled) {
                return;
            }
            if (e.target === this.overAllClear) {
                e.preventDefault();
                return;
            }
            if (!this.inputFocus && this.mode !== 'CheckBox') {
                this.inputElement.focus();
            }
            if (this.mode === 'CheckBox') {
                this.focusIn();
            }
            if (e.target && e.target.classList.toString().indexOf(CHIP_CLOSE) !== -1) {
                if (this.isPopupOpen()) {
                    this.refreshPopup();
                }
                return;
            }
            if (!this.isPopupOpen() && this.openOnClick) {
                this.showPopup();
            }
            else {
                this.hidePopup();
                if (this.mode === 'CheckBox') {
                    this.inputElement.focus();
                }
            }
            e.preventDefault();
        };
        MultiSelect.prototype.enable = function (state) {
            if (state) {
                this.overAllWrapper.classList.remove(DISABLED);
                this.inputElement.removeAttribute('disabled');
                ej2_base_1.attributes(this.inputElement, { 'aria-disabled': 'false' });
            }
            else {
                this.overAllWrapper.classList.add(DISABLED);
                this.inputElement.setAttribute('disabled', 'true');
                ej2_base_1.attributes(this.inputElement, { 'aria-disabled': 'true' });
            }
            if (this.enabled !== state) {
                this.enabled = state;
            }
            this.hidePopup();
        };
        MultiSelect.prototype.onBlur = function (eve) {
            var target = !ej2_base_6.isNullOrUndefined(eve) && eve.relatedTarget;
            if (document.body.contains(this.popupObj.element) && this.popupObj.element.contains(target)) {
                if (this.mode !== 'CheckBox') {
                    this.inputElement.focus();
                }
                return;
            }
            if (this.mode === 'CheckBox' && ej2_base_4.Browser.isIE && !ej2_base_6.isNullOrUndefined(eve)) {
                this.inputFocus = false;
                this.overAllWrapper.classList.remove(FOCUS);
                return;
            }
            if (this.scrollFocusStatus) {
                eve.preventDefault();
                this.inputElement.focus();
                this.scrollFocusStatus = false;
                return;
            }
            this.inputFocus = false;
            this.overAllWrapper.classList.remove(FOCUS);
            if (this.mode !== 'Box' && this.mode !== 'CheckBox') {
                this.refreshListItems(null);
                this.updateDelimView();
            }
            this.updateValueState(eve, this.value, this.tempValues);
            this.dispatchEvent(this.hiddenElement, 'change');
            this.overAllClear.style.display = 'none';
            if (this.isPopupOpen()) {
                this.hidePopup();
            }
            this.makeTextBoxEmpty();
            this.trigger('blur');
            this.focused = true;
            if (ej2_base_4.Browser.isDevice && this.mode !== 'Delimiter' && this.mode !== 'CheckBox') {
                this.removeChipFocus();
            }
            this.removeChipSelection();
            this.refreshInputHight();
            float_label_1.floatLabelBlur(this.overAllWrapper, this.componentWrapper, this.value, this.floatLabelType, this.placeholder);
            this.refreshPlaceHolder();
        };
        MultiSelect.prototype.refreshInputHight = function () {
            if (!this.value || !this.value.length) {
                this.searchWrapper.classList.remove(ZERO_SIZE);
            }
            else {
                this.searchWrapper.classList.add(ZERO_SIZE);
            }
        };
        MultiSelect.prototype.validateValues = function (newValue, oldValue) {
            return JSON.stringify(newValue.slice().sort()) !== JSON.stringify(oldValue.slice().sort());
        };
        MultiSelect.prototype.updateValueState = function (event, newVal, oldVal) {
            var newValue = newVal ? newVal : [];
            var oldValue = oldVal ? oldVal : [];
            if (this.validateValues(newValue, oldValue)) {
                var eventArgs = {
                    e: event,
                    oldValue: oldVal,
                    value: newVal,
                    isInteracted: event ? true : false
                };
                this.trigger('change', eventArgs);
            }
        };
        MultiSelect.prototype.getPagingCount = function () {
            var height = this.list.classList.contains(drop_down_base_1.dropDownBaseClasses.noData) ? null :
                getComputedStyle(this.getItems()[0], null).getPropertyValue('height');
            return Math.round(this.list.getBoundingClientRect().height / parseInt(height, 10));
        };
        MultiSelect.prototype.pageUpSelection = function (steps) {
            var collection = this.list.querySelectorAll('li.'
                + drop_down_base_1.dropDownBaseClasses.li + ':not(.e-reorder-hide)');
            var previousItem;
            if (this.mode === 'CheckBox') {
                previousItem = steps >= 0 ? collection[steps + 1] : collection[0];
            }
            else {
                previousItem = steps >= 0 ? this.liCollections[steps + 1] : this.liCollections[0];
            }
            this.addListFocus(previousItem);
            this.scrollBottom(previousItem, this.getIndexByValue(previousItem.getAttribute('data-value')));
        };
        ;
        MultiSelect.prototype.pageDownSelection = function (steps) {
            var list = this.getItems();
            var collection = this.list.querySelectorAll('li.'
                + drop_down_base_1.dropDownBaseClasses.li + ':not(.e-reorder-hide)');
            var previousItem;
            if (this.mode === 'CheckBox') {
                previousItem = steps <= collection.length ? collection[steps - 1] : collection[collection.length - 1];
            }
            else {
                previousItem = steps <= list.length ? this.liCollections[steps - 1] : this.liCollections[list.length - 1];
            }
            this.addListFocus(previousItem);
            this.scrollBottom(previousItem, this.getIndexByValue(previousItem.getAttribute('data-value')));
        };
        MultiSelect.prototype.focusIn = function () {
            if (this.enabled && !this.readonly) {
                this.inputFocus = true;
                if (!this.value) {
                    this.tempValues = this.value;
                }
                else {
                    this.tempValues = this.value.slice();
                }
                if (this.value && this.value.length) {
                    if (this.mode !== 'Delimiter' && this.mode !== 'CheckBox') {
                        this.chipCollectionWrapper.style.display = '';
                    }
                    else {
                        this.showDelimWrapper();
                    }
                    if (this.mode !== 'CheckBox') {
                        this.viewWrapper.style.display = 'none';
                    }
                }
                if (this.mode !== 'CheckBox') {
                    this.searchWrapper.classList.remove(ZERO_SIZE);
                }
                if (this.focused) {
                    this.inputElement.focus();
                    this.trigger('focus');
                    this.focused = false;
                }
                if (!this.overAllWrapper.classList.contains(FOCUS)) {
                    this.overAllWrapper.classList.add(FOCUS);
                }
                float_label_1.floatLabelFocus(this.overAllWrapper, this.componentWrapper);
                if (this.isPopupOpen()) {
                    this.refreshPopup();
                }
                return true;
            }
            else {
                return false;
            }
        };
        MultiSelect.prototype.showDelimWrapper = function () {
            if (this.mode === 'CheckBox') {
                this.viewWrapper.style.display = '';
            }
            else {
                this.delimiterWrapper.style.display = '';
            }
            this.componentWrapper.classList.add(DELIMITER_VIEW_WRAPPER);
        };
        MultiSelect.prototype.hideDelimWrapper = function () {
            this.delimiterWrapper.style.display = 'none';
            this.componentWrapper.classList.remove(DELIMITER_VIEW_WRAPPER);
        };
        MultiSelect.prototype.expandTextbox = function () {
            var size = 5;
            if (this.placeholder) {
                size = size > this.inputElement.placeholder.length ? size : this.inputElement.placeholder.length;
            }
            if (this.inputElement.value.length > size) {
                this.inputElement.size = this.inputElement.value.length;
            }
            else {
                this.inputElement.size = size;
            }
        };
        MultiSelect.prototype.isPopupOpen = function () {
            return ((this.popupWrapper !== null) && (this.popupWrapper.parentElement !== null));
        };
        MultiSelect.prototype.refreshPopup = function () {
            if (this.popupObj && this.mobFilter) {
                this.popupObj.setProperties({ width: this.calcPopupWidth() });
                this.popupObj.refreshPosition(this.overAllWrapper);
                if (!ej2_base_4.Browser.isIE) {
                    this.popupObj.show();
                }
            }
        };
        MultiSelect.prototype.checkTextLength = function () {
            return this.targetElement().length < 1;
        };
        MultiSelect.prototype.popupKeyActions = function (keyCode) {
            switch (keyCode) {
                case 38:
                    this.hidePopup();
                    if (this.mode === 'CheckBox') {
                        this.inputElement.focus();
                    }
                    break;
                case 40:
                    if (!this.isPopupOpen()) {
                        this.showPopup();
                    }
                    break;
            }
        };
        MultiSelect.prototype.updateAriaAttribute = function () {
            var focusedItem = this.list.querySelector('.' + drop_down_base_1.dropDownBaseClasses.focus);
            if (!ej2_base_6.isNullOrUndefined(focusedItem)) {
                this.inputElement.setAttribute('aria-activedescendant', focusedItem.id);
            }
        };
        MultiSelect.prototype.onKeyDown = function (e) {
            this.keyDownStatus = true;
            if (e.keyCode > 111 && e.keyCode < 124) {
                return;
            }
            if (e.altKey) {
                this.popupKeyActions(e.keyCode);
                e.preventDefault();
                return;
            }
            else if (this.isPopupOpen()) {
                var focusedItem = this.list.querySelector('.' + drop_down_base_1.dropDownBaseClasses.focus);
                var activeIndex = void 0;
                switch (e.keyCode) {
                    case 36:
                    case 35: break;
                    case 33:
                        e.preventDefault();
                        if (focusedItem) {
                            this.getIndexByValue(focusedItem.getAttribute('data-value'));
                            this.pageUpSelection(activeIndex - this.getPagingCount());
                            this.updateAriaAttribute();
                        }
                        return;
                    case 34:
                        e.preventDefault();
                        if (focusedItem) {
                            this.getIndexByValue(focusedItem.getAttribute('data-value'));
                            this.pageDownSelection(activeIndex + this.getPagingCount());
                            this.updateAriaAttribute();
                        }
                        return;
                    case 38:
                        e.preventDefault();
                        this.moveByList(-1);
                        this.updateAriaAttribute();
                        break;
                    case 40:
                        e.preventDefault();
                        this.moveByList(1);
                        this.updateAriaAttribute();
                        break;
                    case 27:
                        e.preventDefault();
                        this.hidePopup();
                        if (this.mode === 'CheckBox') {
                            this.inputElement.focus();
                        }
                        return;
                    case 13:
                        e.preventDefault();
                        if (this.mode !== 'CheckBox') {
                            this.selectByKey(e);
                        }
                        return;
                    case 32:
                        if (this.mode === 'CheckBox') {
                            e.preventDefault();
                            this.selectByKey(e);
                        }
                        return;
                }
            }
            else {
                switch (e.keyCode) {
                    case 13:
                    case 9:
                    case 16:
                    case 17:
                    case 20:
                        return;
                    case 40:
                        if (this.openOnClick) {
                            this.showPopup();
                        }
                        break;
                    case 27:
                        e.preventDefault();
                        this.escapeAction();
                        return;
                }
            }
            if (this.checkTextLength()) {
                if ((this.mode !== 'Delimiter' && this.mode !== 'CheckBox') && this.value && this.value.length) {
                    switch (e.keyCode) {
                        case 37:
                            e.preventDefault();
                            this.moveBy(-1);
                            break;
                        case 39:
                            e.preventDefault();
                            this.moveBy(1);
                            break;
                        case 8:
                            this.removelastSelection(e);
                            break;
                        case 46:
                            this.removeSelectedChip(e);
                            break;
                    }
                }
                else if (e.keyCode === 8 && this.mode === 'Delimiter') {
                    if (this.value && this.value.length) {
                        e.preventDefault();
                        var temp = this.value[this.value.length - 1];
                        this.removeValue(temp, e);
                        this.updateDelimeter(this.delimiterChar);
                        this.focusAtLastListItem(temp);
                    }
                }
            }
            this.expandTextbox();
            this.refreshPopup();
        };
        MultiSelect.prototype.selectByKey = function (e) {
            this.removeChipSelection();
            this.selectListByKey(e);
            if (this.hideSelectedItem) {
                this.focusAtFirstListItem();
            }
        };
        MultiSelect.prototype.escapeAction = function () {
            var temp = this.tempValues ? this.tempValues.slice() : [];
            if (this.value && this.validateValues(this.value, temp)) {
                this.value = temp;
                this.initialValueUpdate();
                if (this.mode !== 'Delimiter' && this.mode !== 'CheckBox') {
                    this.chipCollectionWrapper.style.display = '';
                }
                else {
                    this.showDelimWrapper();
                }
                this.refreshPlaceHolder();
                if (this.value.length) {
                    this.showOverAllClear();
                }
                else {
                    this.hideOverAllClear();
                }
            }
            this.makeTextBoxEmpty();
        };
        MultiSelect.prototype.scrollBottom = function (selectedLI, activeIndex) {
            var currentOffset = this.list.getBoundingClientRect().height;
            var nextBottom = selectedLI.offsetTop + selectedLI.getBoundingClientRect().height - this.list.scrollTop;
            var nextOffset = this.list.scrollTop + nextBottom - currentOffset;
            var boxRange = (selectedLI.offsetTop + selectedLI.getBoundingClientRect().height - this.list.scrollTop);
            boxRange = this.fields.groupBy && !ej2_base_6.isUndefined(this.fixedHeaderElement) ?
                boxRange - this.fixedHeaderElement.getBoundingClientRect().height : boxRange;
            if (activeIndex === 0) {
                this.list.scrollTop = 0;
            }
            else if (nextBottom > currentOffset) {
                this.list.scrollTop = nextOffset;
            }
            else if (!(boxRange > 0 && this.list.getBoundingClientRect().height > boxRange)) {
                this.list.scrollTop = nextOffset;
            }
        };
        MultiSelect.prototype.scrollTop = function (selectedLI, activeIndex) {
            var nextOffset = selectedLI.offsetTop - this.list.scrollTop;
            var nextBottom = selectedLI.offsetTop + selectedLI.getBoundingClientRect().height - this.list.scrollTop;
            nextOffset = this.fields.groupBy && !ej2_base_6.isUndefined(this.fixedHeaderElement) ?
                nextOffset - this.fixedHeaderElement.getBoundingClientRect().height : nextOffset;
            var boxRange = (selectedLI.offsetTop + selectedLI.getBoundingClientRect().height - this.list.scrollTop);
            if (activeIndex === 0) {
                this.list.scrollTop = 0;
            }
            else if (nextOffset < 0) {
                this.list.scrollTop = this.list.scrollTop + nextOffset;
            }
            else if (!(boxRange > 0 && this.list.getBoundingClientRect().height > boxRange)) {
                this.list.scrollTop = selectedLI.offsetTop - (this.fields.groupBy && !ej2_base_6.isUndefined(this.fixedHeaderElement) ?
                    this.fixedHeaderElement.getBoundingClientRect().height : 0);
            }
        };
        MultiSelect.prototype.selectListByKey = function (e) {
            var li = this.list.querySelector('li.' + drop_down_base_1.dropDownBaseClasses.focus);
            var limit = this.value && this.value.length ? this.value.length : 0;
            if (li !== null) {
                if (li.classList.contains('e-active')) {
                    limit = limit - 1;
                }
                if (this.isValidLI(li) && limit < this.maximumSelectionLength) {
                    this.updateListSelection(li, e);
                    this.addListFocus(li);
                    if (this.mode === 'CheckBox') {
                        this.updateDelimView();
                        this.refreshInputHight();
                        this.updateDelimeter(this.delimiterChar);
                    }
                    else {
                        this.updateDelimeter(this.delimiterChar);
                    }
                    this.makeTextBoxEmpty();
                    if (this.mode !== 'CheckBox') {
                        this.refreshListItems(li.textContent);
                    }
                    this.refreshPopup();
                    if (this.value && this.value.length) {
                        this.removeListSelection();
                        this.refreshSelection();
                    }
                }
                if (this.closePopupOnSelect) {
                    this.hidePopup();
                }
            }
            this.refreshPlaceHolder();
        };
        MultiSelect.prototype.refreshListItems = function (data) {
            if ((this.allowFiltering || this.allowCustomValue) && this.mainList && this.listData) {
                var list = this.mainList.cloneNode ? this.mainList.cloneNode(true) : this.mainList;
                this.onActionComplete(list, this.mainData);
                this.focusAtLastListItem(data);
                if (this.value && this.value.length) {
                    this.refreshSelection();
                }
            }
        };
        MultiSelect.prototype.removeSelectedChip = function (e) {
            var selectedElem = this.chipCollectionWrapper.querySelector('span.' + CHIP_SELECTED);
            var temp;
            if (selectedElem !== null) {
                if (!ej2_base_6.isNullOrUndefined(this.value)) {
                    this.tempValues = this.value.slice();
                }
                temp = selectedElem.nextElementSibling;
                if (temp !== null) {
                    this.removeChipSelection();
                    this.addChipSelection(temp);
                }
                this.removeValue(selectedElem.getAttribute('data-value'), e);
                this.makeTextBoxEmpty();
            }
            if (this.closePopupOnSelect) {
                this.hidePopup();
            }
        };
        MultiSelect.prototype.moveByTop = function (state) {
            var elements = this.list.querySelectorAll('li.' + drop_down_base_1.dropDownBaseClasses.li);
            var index;
            if (elements.length > 1) {
                this.removeFocus();
                index = state ? 0 : (elements.length - 1);
                this.addListFocus(elements[index]);
                this.scrollBottom(elements[index], index);
            }
            this.updateAriaAttribute();
        };
        MultiSelect.prototype.moveByList = function (position) {
            if (this.list) {
                var elements = this.list.querySelectorAll('li.'
                    + drop_down_base_1.dropDownBaseClasses.li
                    + ':not(.' + HIDE_LIST + ')' + ':not(.e-reorder-hide)');
                var selectedElem = this.list.querySelector('li.' + drop_down_base_1.dropDownBaseClasses.focus);
                var temp = -1;
                if (elements.length) {
                    for (var index = 0; index < elements.length; index++) {
                        if (elements[index] === selectedElem) {
                            temp = index;
                            break;
                        }
                    }
                    if (position > 0) {
                        if (temp < (elements.length - 1)) {
                            this.removeFocus();
                            this.addListFocus(elements[++temp]);
                            this.scrollBottom(elements[temp], temp);
                        }
                    }
                    else {
                        if (temp > 0) {
                            this.removeFocus();
                            this.addListFocus(elements[--temp]);
                            this.scrollTop(elements[temp], temp);
                        }
                    }
                }
            }
        };
        MultiSelect.prototype.moveBy = function (position) {
            var elements;
            var selectedElem;
            var temp;
            elements = this.chipCollectionWrapper.querySelectorAll('span.' + CHIP);
            selectedElem = this.chipCollectionWrapper.querySelector('span.' + CHIP_SELECTED);
            if (selectedElem === null) {
                if (position < 0) {
                    this.addChipSelection(elements[elements.length - 1]);
                }
            }
            else {
                if (position < 0) {
                    temp = selectedElem.previousElementSibling;
                    if (temp !== null) {
                        this.removeChipSelection();
                        this.addChipSelection(temp);
                    }
                }
                else {
                    temp = selectedElem.nextElementSibling;
                    this.removeChipSelection();
                    if (temp !== null) {
                        this.addChipSelection(temp);
                    }
                }
            }
        };
        MultiSelect.prototype.chipClick = function (e) {
            if (this.enabled) {
                var elem = ej2_base_5.closest(e.target, '.' + CHIP);
                this.removeChipSelection();
                this.addChipSelection(elem);
                this.trigger('chipSelection', e);
            }
        };
        MultiSelect.prototype.removeChipSelection = function () {
            if (this.chipCollectionWrapper) {
                this.removeChipFocus();
            }
        };
        MultiSelect.prototype.addChipSelection = function (element) {
            ej2_base_5.addClass([element], CHIP_SELECTED);
        };
        MultiSelect.prototype.onChipRemove = function (e) {
            if (e.which === 3 || e.button === 2) {
                return;
            }
            if (this.enabled && !this.readonly) {
                var element = e.target.parentElement;
                var value = this.getFormattedValue(element.getAttribute('data-value'));
                if (this.isPopupOpen() && this.mode !== 'CheckBox') {
                    this.hidePopup();
                }
                if (!this.inputFocus) {
                    this.inputElement.focus();
                }
                this.removeValue(value, e);
                this.updateDelimeter(this.delimiterChar);
                this.makeTextBoxEmpty();
                e.preventDefault();
            }
        };
        MultiSelect.prototype.makeTextBoxEmpty = function () {
            this.inputElement.value = '';
            this.refreshPlaceHolder();
        };
        MultiSelect.prototype.refreshPlaceHolder = function () {
            if (this.placeholder && this.floatLabelType === 'Never') {
                if (this.value && this.value.length) {
                    this.inputElement.placeholder = '';
                }
                else {
                    this.inputElement.placeholder = this.placeholder;
                }
            }
            else {
                this.setFloatLabelType();
            }
            this.expandTextbox();
        };
        MultiSelect.prototype.removeValue = function (value, eve) {
            var index = this.value.indexOf(this.getFormattedValue(value));
            var className = this.hideSelectedItem ?
                HIDE_LIST :
                drop_down_base_1.dropDownBaseClasses.selected;
            if (index !== -1) {
                var element = this.list.querySelector('li[data-value="' + value + '"]');
                var eventArgs = {
                    e: eve,
                    item: element,
                    itemData: this.getDataByValue(value),
                    isInteracted: eve ? true : false,
                    cancel: false
                };
                this.trigger('removing', eventArgs);
                if (eventArgs.cancel) {
                    return;
                }
                this.value.splice(index, 1);
                this.setProperties({ value: [].concat([], this.value) }, true);
                if (element !== null) {
                    element.setAttribute('aria-selected', 'false');
                    ej2_base_5.removeClass([element], className);
                    this.notify('activeList', {
                        module: 'CheckBoxSelection',
                        enable: this.mode === 'CheckBox', li: element,
                        e: this, index: index
                    });
                    this.notify('updatelist', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', li: element, e: eve });
                    ej2_base_1.attributes(this.inputElement, { 'aria-activedescendant': element.id });
                    if ((this.value.length !== this.mainData.length) && (this.mode === 'CheckBox' && this.showSelectAll)) {
                        this.notify('checkSelectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', value: 'uncheck' });
                    }
                }
                this.updateMainList(true, value);
                this.removeChip(value);
                this.updateChipStatus();
                var limit = this.value && this.value.length ? this.value.length : 0;
                if (limit < this.maximumSelectionLength) {
                    var collection = this.list.querySelectorAll('li.'
                        + drop_down_base_1.dropDownBaseClasses.li + ':not(.e-active)');
                    ej2_base_5.removeClass(collection, 'e-disable');
                }
                this.trigger('removed', eventArgs);
            }
        };
        MultiSelect.prototype.updateMainList = function (state, value) {
            if (this.allowFiltering) {
                var element2 = this.mainList.querySelector('li[data-value="' + value + '"]');
                if (element2) {
                    if (state) {
                        element2.setAttribute('aria-selected', 'false');
                        ej2_base_5.removeClass([element2], this.hideSelectedItem ?
                            HIDE_LIST :
                            drop_down_base_1.dropDownBaseClasses.selected);
                        if (this.mode === 'CheckBox') {
                            element2.firstElementChild.setAttribute('aria-checked', 'false');
                            ej2_base_5.removeClass([element2.firstElementChild.lastElementChild], 'e-check');
                        }
                    }
                    else {
                        element2.setAttribute('aria-selected', 'true');
                        ej2_base_5.addClass([element2], this.hideSelectedItem ?
                            HIDE_LIST :
                            drop_down_base_1.dropDownBaseClasses.selected);
                        if (this.mode === 'CheckBox') {
                            element2.firstElementChild.setAttribute('aria-checked', 'true');
                            ej2_base_5.addClass([element2.firstElementChild.lastElementChild], 'e-check');
                        }
                    }
                }
            }
        };
        MultiSelect.prototype.removeChip = function (value) {
            if (this.chipCollectionWrapper) {
                var element = this.chipCollectionWrapper.querySelector('span[data-value="' + value + '"]');
                if (element) {
                    ej2_base_5.remove(element);
                }
            }
        };
        MultiSelect.prototype.updateChipStatus = function () {
            if (this.value.length) {
                if (!ej2_base_6.isNullOrUndefined(this.chipCollectionWrapper)) {
                    (this.chipCollectionWrapper.style.display = '');
                }
                if (this.mode === 'Delimiter' || this.mode === 'CheckBox') {
                    this.showDelimWrapper();
                }
                this.showOverAllClear();
            }
            else {
                if (!ej2_base_6.isNullOrUndefined(this.chipCollectionWrapper)) {
                    this.chipCollectionWrapper.style.display = 'none';
                }
                if (!ej2_base_6.isNullOrUndefined(this.delimiterWrapper)) {
                    (this.delimiterWrapper.style.display = 'none');
                }
                this.hideOverAllClear();
            }
        };
        MultiSelect.prototype.addValue = function (value, text, eve) {
            if (!this.value) {
                this.value = [];
            }
            this.setProperties({ value: [].concat([], this.value, [value]) }, true);
            var element = this.list.querySelector('li[data-value="' + value + '"]');
            this.removeFocus();
            if (element) {
                this.addListFocus(element);
                this.addListSelection(element);
            }
            if (this.mode !== 'Delimiter' && this.mode !== 'CheckBox') {
                this.addChip(text, value, eve);
            }
            this.updateChipStatus();
            this.checkMaxSelection();
        };
        MultiSelect.prototype.checkMaxSelection = function () {
            var limit = this.value && this.value.length ? this.value.length : 0;
            if (limit === this.maximumSelectionLength) {
                var collection = this.list.querySelectorAll('li.'
                    + drop_down_base_1.dropDownBaseClasses.li + ':not(.e-active)');
                ej2_base_5.addClass(collection, 'e-disable');
            }
        };
        MultiSelect.prototype.dispatchSelect = function (value, eve, element, isNotTrigger) {
            if (this.initStatus && !isNotTrigger) {
                var eventArgs = {
                    e: eve,
                    item: element,
                    itemData: this.getDataByValue(value),
                    isInteracted: eve ? true : false,
                    cancel: false
                };
                this.trigger('select', eventArgs);
                if (eventArgs.cancel) {
                    return true;
                }
            }
            return false;
        };
        MultiSelect.prototype.addChip = function (text, value, e) {
            if (this.chipCollectionWrapper) {
                var item = this.getChip(text, value, e);
                if (item.cancel) {
                    return;
                }
                this.chipCollectionWrapper.appendChild(item.element);
            }
        };
        MultiSelect.prototype.removeChipFocus = function () {
            var elements;
            var closeElements;
            elements = this.chipCollectionWrapper.querySelectorAll('span.' + CHIP);
            closeElements = this.chipCollectionWrapper.querySelectorAll('span.' + CHIP_CLOSE.split(' ')[0]);
            ej2_base_5.removeClass(elements, CHIP_SELECTED);
            if (ej2_base_4.Browser.isDevice) {
                for (var index = 0; index < closeElements.length; index++) {
                    closeElements[index].style.display = 'none';
                }
            }
        };
        MultiSelect.prototype.onMobileChipInteraction = function (e) {
            var chipElem = ej2_base_5.closest(e.target, '.' + CHIP);
            var chipClose = chipElem.querySelector('span.' + CHIP_CLOSE.split(' ')[0]);
            if (this.enabled && !this.readonly) {
                if (!chipElem.classList.contains(CHIP_SELECTED)) {
                    this.removeChipFocus();
                    chipClose.style.display = '';
                    chipElem.classList.add(CHIP_SELECTED);
                }
                this.refreshPopup();
                e.preventDefault();
            }
        };
        MultiSelect.prototype.getChip = function (data, value, e) {
            var itemData = { text: value, value: value };
            var chip = ej2_base_5.createElement('span', {
                className: CHIP,
                attrs: { 'data-value': value, 'title': data }
            });
            var chipContent = ej2_base_5.createElement('span', { className: CHIP_CONTENT });
            var chipClose = ej2_base_5.createElement('span', { className: CHIP_CLOSE });
            if (this.mainData) {
                itemData = this.getDataByValue(value);
            }
            if (this.valueTemplate && itemData) {
                var compiledString = ej2_base_3.compile(this.valueTemplate);
                for (var _i = 0, _a = compiledString(itemData); _i < _a.length; _i++) {
                    var item = _a[_i];
                    chipContent.appendChild(item);
                }
            }
            else {
                chipContent.innerHTML = data;
            }
            chip.appendChild(chipContent);
            var eventArgs = {
                isInteracted: e ? true : false,
                itemData: itemData,
                e: e,
                setClass: function (classes) {
                    ej2_base_5.addClass([chip], classes);
                },
                cancel: false
            };
            this.trigger('tagging', eventArgs);
            if (eventArgs.cancel) {
                return { cancel: true, element: chip };
            }
            if (ej2_base_4.Browser.isDevice) {
                chip.classList.add(MOBILE_CHIP);
                ej2_base_5.append([chipClose], chip);
                chipClose.style.display = 'none';
                ej2_base_3.EventHandler.add(chip, 'click', this.onMobileChipInteraction, this);
            }
            else {
                ej2_base_3.EventHandler.add(chip, 'mousedown', this.chipClick, this);
                if (this.showClearButton) {
                    chip.appendChild(chipClose);
                }
            }
            ej2_base_3.EventHandler.add(chipClose, 'mousedown', this.onChipRemove, this);
            return { cancel: false, element: chip };
        };
        MultiSelect.prototype.calcPopupWidth = function () {
            var width = ej2_base_6.formatUnit(this.popupWidth);
            if (width.indexOf('%') > -1) {
                var inputWidth = (this.componentWrapper.getBoundingClientRect().width) * parseFloat(width) / 100;
                width = inputWidth.toString() + 'px';
            }
            return width;
        };
        MultiSelect.prototype.mouseIn = function () {
            if (this.enabled && !this.readonly) {
                this.showOverAllClear();
            }
        };
        MultiSelect.prototype.mouseOut = function () {
            if (!this.inputFocus) {
                this.overAllClear.style.display = 'none';
            }
        };
        MultiSelect.prototype.listOption = function (dataSource, fields) {
            var iconCss = ej2_base_6.isNullOrUndefined(fields.iconCss) ? false : true;
            var fieldProperty = fields.properties;
            this.listCurrentOptions = (fields.text !== null || fields.value !== null) ? {
                fields: fieldProperty, showIcon: iconCss, ariaAttributes: { groupItemRole: 'presentation' }
            } : { fields: { value: 'text' } };
            ej2_base_2.extend(this.listCurrentOptions, this.listCurrentOptions, fields, true);
            if (this.mode === 'CheckBox') {
                this.notify('listoption', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', dataSource: dataSource, fieldProperty: fieldProperty });
            }
            return this.listCurrentOptions;
        };
        MultiSelect.prototype.renderPopup = function () {
            var _this = this;
            if (!this.list) {
                _super.prototype.render.call(this);
            }
            if (!this.popupObj) {
                document.body.appendChild(this.popupWrapper);
                var overAllHeight = parseInt(this.popupHeight, 10);
                this.popupWrapper.style.visibility = 'hidden';
                if (this.headerTemplate) {
                    var compiledString = void 0;
                    this.header = document.createElement('div');
                    ej2_base_5.addClass([this.header], HEADER);
                    compiledString = ej2_base_3.compile(this.headerTemplate);
                    var elements = compiledString({});
                    for (var temp = 0; temp < elements.length; temp++) {
                        this.header.appendChild(elements[temp]);
                    }
                    if (this.mode === 'CheckBox' && this.showSelectAll) {
                        ej2_base_4.prepend([this.header], this.popupWrapper);
                    }
                    else {
                        ej2_base_5.append([this.header], this.popupWrapper);
                    }
                    ej2_base_3.EventHandler.add(this.header, 'mousedown', this.onListMouseDown, this);
                    overAllHeight -= this.header.getBoundingClientRect().height;
                }
                ej2_base_5.append([this.list], this.popupWrapper);
                if (this.footerTemplate) {
                    var compiledString = void 0;
                    this.footer = document.createElement('div');
                    ej2_base_5.addClass([this.footer], FOOTER);
                    compiledString = ej2_base_3.compile(this.footerTemplate);
                    var elements = compiledString({});
                    for (var temp = 0; temp < elements.length; temp++) {
                        this.footer.appendChild(elements[temp]);
                    }
                    ej2_base_5.append([this.footer], this.popupWrapper);
                    ej2_base_3.EventHandler.add(this.footer, 'mousedown', this.onListMouseDown, this);
                    overAllHeight -= this.footer.getBoundingClientRect().height;
                }
                if (this.mode === 'CheckBox' && this.showSelectAll) {
                    this.notify('selectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox' });
                    overAllHeight -= this.selectAllHeight;
                }
                else if (this.mode === 'CheckBox' && !this.showSelectAll) {
                    this.notify('selectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox' });
                    overAllHeight = parseInt(this.popupHeight, 10);
                }
                if (this.mode === 'CheckBox') {
                    var args = {
                        module: 'CheckBoxSelection',
                        enable: this.mode === 'CheckBox',
                        popupElement: this.popupWrapper
                    };
                    this.notify('searchBox', args);
                    overAllHeight -= this.searchBoxHeight;
                    ej2_base_5.addClass([this.popupWrapper], 'e-checkbox');
                }
                if (this.popupHeight !== 'auto') {
                    this.list.style.maxHeight = ej2_base_6.formatUnit(overAllHeight);
                    this.popupWrapper.style.maxHeight = ej2_base_6.formatUnit(this.popupHeight);
                }
                else {
                    this.list.style.maxHeight = ej2_base_6.formatUnit(this.popupHeight);
                }
                this.popupObj = new ej2_popups_1.Popup(this.popupWrapper, {
                    width: this.calcPopupWidth(), targetType: 'relative', position: { X: 'left', Y: 'bottom' },
                    relateTo: this.overAllWrapper, collision: { X: 'flip', Y: 'flip' }, offsetY: 1,
                    enableRtl: this.enableRtl,
                    zIndex: this.zIndex,
                    close: function () {
                        if (_this.popupObj.element.parentElement) {
                            ej2_base_5.detach(_this.popupObj.element);
                        }
                    },
                    open: function () {
                        _this.notify('inputFocus', { module: 'CheckBoxSelection', enable: _this.mode === 'CheckBox', value: 'focus' });
                    }
                });
                this.popupObj.close();
                this.popupWrapper.style.visibility = '';
                if (this.mode === 'CheckBox' && ej2_base_4.Browser.isDevice) {
                    this.notify('deviceSearchBox', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox' });
                }
            }
        };
        MultiSelect.prototype.ClearAll = function (e) {
            if (this.enabled && !this.readonly) {
                var temp = void 0;
                var tempValues = this.value ? this.value.slice() : [];
                for (temp = this.value[0]; this.value.length !== 0; temp = this.value[0]) {
                    this.removeValue(temp, e);
                }
                this.updateDelimeter(this.delimiterChar);
                if (this.mode !== 'Box') {
                    this.updateDelimView();
                }
                this.makeTextBoxEmpty();
                if (this.isPopupOpen()) {
                    this.refreshPopup();
                }
                this.removeFocus();
                if (!this.inputFocus) {
                    this.updateValueState(e, this.value, tempValues);
                    if (this.mode !== 'CheckBox') {
                        this.inputElement.focus();
                    }
                }
                if (this.mode === 'CheckBox') {
                    this.refreshPlaceHolder();
                    this.refreshInputHight();
                }
                e.preventDefault();
            }
        };
        MultiSelect.prototype.windowResize = function () {
            this.refreshPopup();
            if (!this.inputFocus && this.viewWrapper && this.viewWrapper.parentElement) {
                this.updateDelimView();
            }
        };
        MultiSelect.prototype.wireEvent = function () {
            ej2_base_3.EventHandler.add(this.componentWrapper, 'mousedown', this.wrapperClick, this);
            ej2_base_3.EventHandler.add(window, 'resize', this.windowResize, this);
            ej2_base_3.EventHandler.add(this.inputElement, 'focus', this.focusIn, this);
            ej2_base_3.EventHandler.add(this.inputElement, 'keydown', this.onKeyDown, this);
            ej2_base_3.EventHandler.add(this.inputElement, 'keyup', this.KeyUp, this);
            ej2_base_3.EventHandler.add(this.inputElement, 'input', this.onInput, this);
            ej2_base_3.EventHandler.add(this.inputElement, 'blur', this.onBlur, this);
            ej2_base_3.EventHandler.add(this.componentWrapper, 'mousemove', this.mouseIn, this);
            ej2_base_3.EventHandler.add(this.componentWrapper, 'mouseout', this.mouseOut, this);
            ej2_base_3.EventHandler.add(this.overAllClear, 'mouseup', this.ClearAll, this);
        };
        MultiSelect.prototype.onInput = function () {
            if (this.keyDownStatus) {
                this.isValidKey = true;
            }
            else {
                this.isValidKey = false;
            }
            this.keyDownStatus = false;
        };
        MultiSelect.prototype.preRender = function () {
            this.initializeData();
            _super.prototype.preRender.call(this);
        };
        MultiSelect.prototype.initializeData = function () {
            this.mainListCollection = [];
            this.beforePopupOpen = false;
            this.filterAction = false;
            this.remoteFilterAction = false;
            this.isFirstClick = false;
        };
        MultiSelect.prototype.updateData = function (delimiterChar) {
            var data = '';
            var delim = this.mode === 'Delimiter' || this.mode === 'CheckBox';
            var text = [];
            var temp;
            var tempData = this.listData;
            this.listData = this.mainData;
            this.hiddenElement.innerHTML = '';
            if (!ej2_base_6.isNullOrUndefined(this.value)) {
                for (var index = 0; this.value[index]; index++) {
                    if (this.listData) {
                        temp = this.getTextByValue(this.value[index]);
                    }
                    else {
                        temp = this.value[index];
                    }
                    data += temp + delimiterChar + ' ';
                    text.push(temp);
                    this.hiddenElement.innerHTML += '<option selected value =' + this.value[index] + '>' + index + '</option>';
                }
            }
            this.text = text.toString();
            if (delim) {
                this.delimiterWrapper.innerHTML = data;
            }
            this.listData = tempData;
        };
        MultiSelect.prototype.initialValueUpdate = function () {
            if (this.list) {
                var text = void 0;
                var textField = void 0;
                var valueField = void 0;
                var element = void 0;
                var value = void 0;
                if (this.chipCollectionWrapper) {
                    this.chipCollectionWrapper.innerHTML = '';
                }
                this.removeListSelection();
                if (!ej2_base_6.isNullOrUndefined(this.value)) {
                    for (var index = 0; this.value[index]; index++) {
                        value = this.value[index];
                        element = this.hideSelectedItem ? this.ulElement.querySelector('li[data-value="' + value + '"]')
                            : this.list.querySelector('li[data-value="' + value + '"]');
                        text = this.getTextByValue(value);
                        if ((element && (element.getAttribute('aria-selected') !== 'true')) ||
                            (element && (element.getAttribute('aria-selected') === 'true' && this.hideSelectedItem) &&
                                (this.mode === 'Box' || this.mode === 'Default'))) {
                            text = this.getTextByValue(value);
                            this.addChip(text, value);
                            this.addListSelection(element);
                        }
                        else if (value && this.allowCustomValue) {
                            var indexItem = this.listData.length;
                            var newValue = {};
                            ej2_base_1.setValue(this.fields.text, value, newValue);
                            ej2_base_1.setValue(this.fields.value, value, newValue);
                            this.addItem(newValue, indexItem);
                            text = this.getTextByValue(value);
                            this.addChip(text, value);
                            this.addListSelection(element);
                        }
                    }
                }
                if (this.mode === 'CheckBox') {
                    this.updateDelimView();
                    this.updateValueState(null, this.value, this.tempValues);
                    this.refreshInputHight();
                    this.updateDelimeter(this.delimiterChar);
                }
                else {
                    this.updateDelimeter(this.delimiterChar);
                }
                if (this.mode === 'CheckBox' && this.showSelectAll && ej2_base_6.isNullOrUndefined(this.value)) {
                    this.notify('checkSelectAll', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', value: 'uncheck' });
                }
                if (!this.inputFocus) {
                    if (this.mode === 'Box') {
                        this.chipCollectionWrapper.style.display = '';
                    }
                    else if (this.mode === 'Delimiter' || this.mode === 'CheckBox') {
                        this.showDelimWrapper();
                    }
                }
            }
        };
        MultiSelect.prototype.isValidLI = function (li) {
            return (li && !li.classList.contains(drop_down_base_1.dropDownBaseClasses.disabled) && !li.classList.contains(drop_down_base_1.dropDownBaseClasses.group) &&
                li.classList.contains(drop_down_base_1.dropDownBaseClasses.li));
        };
        ;
        MultiSelect.prototype.updateListSelection = function (li, e) {
            var value = this.getFormattedValue(li.getAttribute('data-value'));
            var text = this.getTextByValue(value);
            this.removeHover();
            if (!this.value || this.value.indexOf(value) === -1) {
                var argsCancel = this.dispatchSelect(value, e, li, (li.getAttribute('aria-selected') === 'true'));
                if (argsCancel) {
                    return;
                }
                if ((this.allowCustomValue || this.allowFiltering) && !this.mainList.querySelector('li[data-value="' + value + '"]')) {
                    var temp = li.cloneNode(true);
                    var data = this.getDataByValue(value);
                    ej2_base_5.append([temp], this.mainList);
                    this.mainData.push(this.getDataByValue(value));
                    var eventArgs = {
                        newData: data,
                        cancel: false
                    };
                    this.trigger('customValueSelection', eventArgs);
                    if (eventArgs.cancel) {
                        return;
                    }
                }
                this.remoteCustomValue = false;
                this.addValue(value, text, e);
            }
            else {
                this.removeValue(value, e);
            }
        };
        MultiSelect.prototype.removeListSelection = function () {
            var className = this.hideSelectedItem ?
                HIDE_LIST :
                drop_down_base_1.dropDownBaseClasses.selected;
            var selectedItems = this.list.querySelectorAll('.' + className);
            var temp = selectedItems.length;
            if (selectedItems && selectedItems.length) {
                ej2_base_5.removeClass(selectedItems, className);
                while (temp > 0) {
                    selectedItems[temp - 1].setAttribute('aria-selected', 'false');
                    temp--;
                }
            }
            if (!ej2_base_6.isNullOrUndefined(this.mainList)) {
                var selectItems = this.mainList.querySelectorAll('.' + className);
                var temp1 = selectItems.length;
                if (selectItems && selectItems.length) {
                    ej2_base_5.removeClass(selectItems, className);
                    while (temp1 > 0) {
                        selectItems[temp1 - 1].setAttribute('aria-selected', 'false');
                        if (this.mode === 'CheckBox') {
                            selectItems[temp1 - 1].firstElementChild.setAttribute('aria-checked', 'false');
                            ej2_base_5.removeClass([selectItems[temp1 - 1].firstElementChild.lastElementChild], 'e-check');
                        }
                        temp1--;
                    }
                }
            }
        };
        ;
        MultiSelect.prototype.removeHover = function () {
            var hoveredItem = this.list.querySelectorAll('.' + drop_down_base_1.dropDownBaseClasses.hover);
            if (hoveredItem && hoveredItem.length) {
                ej2_base_5.removeClass(hoveredItem, drop_down_base_1.dropDownBaseClasses.hover);
            }
        };
        ;
        MultiSelect.prototype.removeFocus = function () {
            var hoveredItem = this.list.querySelectorAll('.' + drop_down_base_1.dropDownBaseClasses.focus);
            var mainlist = this.mainList.querySelectorAll('.' + drop_down_base_1.dropDownBaseClasses.focus);
            if (hoveredItem && hoveredItem.length) {
                ej2_base_5.removeClass(hoveredItem, drop_down_base_1.dropDownBaseClasses.focus);
                ej2_base_5.removeClass(mainlist, drop_down_base_1.dropDownBaseClasses.focus);
            }
        };
        ;
        MultiSelect.prototype.addListHover = function (li) {
            if (this.enabled && this.isValidLI(li)) {
                this.removeHover();
                ej2_base_5.addClass([li], drop_down_base_1.dropDownBaseClasses.hover);
            }
        };
        ;
        MultiSelect.prototype.addListFocus = function (element) {
            if (this.enabled && this.isValidLI(element)) {
                this.removeFocus();
                ej2_base_5.addClass([element], drop_down_base_1.dropDownBaseClasses.focus);
            }
        };
        MultiSelect.prototype.addListSelection = function (element) {
            var className = this.hideSelectedItem ?
                HIDE_LIST :
                drop_down_base_1.dropDownBaseClasses.selected;
            if (this.isValidLI(element) && !element.classList.contains(drop_down_base_1.dropDownBaseClasses.hover)) {
                ej2_base_5.addClass([element], className);
                this.updateMainList(false, element.getAttribute('data-value'));
                element.setAttribute('aria-selected', 'true');
                if (this.mode === 'CheckBox') {
                    var ariaCheck = element.firstElementChild.getAttribute('aria-checked');
                    if (ariaCheck === 'false' || ej2_base_6.isNullOrUndefined(ariaCheck)) {
                        this.notify('updatelist', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', li: element, e: this });
                    }
                }
                this.notify('activeList', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', li: element, e: this });
                if (this.chipCollectionWrapper !== null) {
                    this.removeChipSelection();
                }
                ej2_base_1.attributes(this.inputElement, { 'aria-activedescendant': element.id });
            }
        };
        MultiSelect.prototype.updateDelimeter = function (delimChar) {
            this.updateData(delimChar);
        };
        MultiSelect.prototype.onMouseClick = function (e) {
            this.scrollFocusStatus = false;
            var target = e.target;
            var li = ej2_base_5.closest(target, '.' + drop_down_base_1.dropDownBaseClasses.li);
            if (this.isValidLI(li)) {
                var limit = this.value && this.value.length ? this.value.length : 0;
                if (li.classList.contains('e-active')) {
                    limit = limit - 1;
                }
                if (limit < this.maximumSelectionLength) {
                    this.updateListSelection(li, e);
                    this.addListFocus(li);
                    if ((this.allowCustomValue || this.allowFiltering) && this.mainList && this.listData) {
                        if (this.mode !== 'CheckBox') {
                            this.focusAtLastListItem(li.getAttribute('data-value'));
                        }
                        this.refreshSelection();
                    }
                    else {
                        this.makeTextBoxEmpty();
                    }
                }
                if (this.mode === 'CheckBox') {
                    this.updateDelimView();
                    this.refreshInputHight();
                    this.updateDelimeter(this.delimiterChar);
                }
                else {
                    this.updateDelimeter(this.delimiterChar);
                }
                this.checkSelectAll();
                this.refreshPopup();
                if (this.hideSelectedItem) {
                    this.focusAtFirstListItem();
                }
                if (this.closePopupOnSelect) {
                    this.hidePopup();
                }
                else {
                    e.preventDefault();
                }
                this.makeTextBoxEmpty();
            }
            else {
                e.preventDefault();
            }
            if (this.mode !== 'CheckBox') {
                this.refreshListItems(ej2_base_6.isNullOrUndefined(li) ? null : li.textContent);
            }
            this.refreshPlaceHolder();
        };
        MultiSelect.prototype.onMouseOver = function (e) {
            var currentLi = ej2_base_5.closest(e.target, '.' + drop_down_base_1.dropDownBaseClasses.li);
            this.addListHover(currentLi);
        };
        MultiSelect.prototype.onMouseLeave = function (e) {
            this.removeHover();
        };
        MultiSelect.prototype.onListMouseDown = function (e) {
            e.preventDefault();
            this.scrollFocusStatus = true;
        };
        MultiSelect.prototype.wireListEvents = function () {
            ej2_base_3.EventHandler.add(this.list, 'mousedown', this.onListMouseDown, this);
            ej2_base_3.EventHandler.add(this.list, 'mouseup', this.onMouseClick, this);
            ej2_base_3.EventHandler.add(this.list, 'mouseover', this.onMouseOver, this);
            ej2_base_3.EventHandler.add(this.list, 'mouseout', this.onMouseLeave, this);
        };
        ;
        MultiSelect.prototype.unwireListEvents = function () {
            if (this.list) {
                ej2_base_3.EventHandler.remove(this.list, 'mousedown', this.onListMouseDown);
                ej2_base_3.EventHandler.remove(this.list, 'mouseup', this.onMouseClick);
                ej2_base_3.EventHandler.remove(this.list, 'mouseover', this.onMouseOver);
                ej2_base_3.EventHandler.remove(this.list, 'mouseout', this.onMouseLeave);
            }
        };
        ;
        MultiSelect.prototype.hideOverAllClear = function () {
            if (!this.value || !this.value.length) {
                this.overAllClear.style.display = 'none';
            }
        };
        MultiSelect.prototype.showOverAllClear = function () {
            if (this.value && this.value.length) {
                this.overAllClear.style.display = '';
            }
        };
        MultiSelect.prototype.showSpinner = function () {
            if (ej2_base_6.isNullOrUndefined(this.spinnerElement)) {
                if (this.overAllClear.style.display !== 'none') {
                    this.spinnerElement = this.overAllClear;
                }
                else {
                    this.spinnerElement = ej2_base_5.createElement('span', { className: CLOSEICON_CLASS + ' ' + SPINNER_CLASS });
                    this.componentWrapper.appendChild(this.spinnerElement);
                }
                ej2_base_5.addClass([this.spinnerElement], DISABLE_ICON);
                ej2_popups_1.createSpinner({
                    target: this.spinnerElement,
                    width: ej2_base_4.Browser.isDevice ? '16px' : '14px'
                });
                ej2_popups_1.showSpinner(this.spinnerElement);
            }
        };
        MultiSelect.prototype.hideSpinner = function () {
            if (!ej2_base_6.isNullOrUndefined(this.spinnerElement)) {
                ej2_popups_1.hideSpinner(this.spinnerElement);
                ej2_base_5.removeClass([this.spinnerElement], DISABLE_ICON);
                if (this.spinnerElement.classList.contains(SPINNER_CLASS)) {
                    ej2_base_5.detach(this.spinnerElement);
                }
                else {
                    this.spinnerElement.innerHTML = '';
                }
                this.spinnerElement = null;
            }
        };
        MultiSelect.prototype.updateDelimView = function () {
            if (this.delimiterWrapper) {
                this.hideDelimWrapper();
            }
            if (this.chipCollectionWrapper) {
                this.chipCollectionWrapper.style.display = 'none';
            }
            this.viewWrapper.style.display = '';
            if (this.value && this.value.length) {
                var data = '';
                var temp = void 0;
                var tempData = void 0;
                var tempIndex = 1;
                var wrapperleng = void 0;
                var remaining = void 0;
                this.viewWrapper.innerHTML = '';
                var l10nLocale = {
                    noRecordsTemplate: 'No Records Found',
                    actionFailureTemplate: 'The Request Failed',
                    overflowCountTemplate: '+${count} more..'
                };
                var l10n = new ej2_base_3.L10n('dropdowns', l10nLocale, this.locale);
                var remainContent = l10n.getConstant('overflowCountTemplate');
                var raminElement = ej2_base_5.createElement('span', {
                    className: REMAIN_WRAPPER
                });
                var compiledString = ej2_base_3.compile(remainContent);
                raminElement.appendChild(compiledString({ 'count': this.value.length })[0]);
                this.viewWrapper.appendChild(raminElement);
                var remainSize = raminElement.getBoundingClientRect().width;
                ej2_base_5.remove(raminElement);
                this.viewWrapper.innerHTML = '';
                var inputleng = this.searchWrapper.getBoundingClientRect().width;
                var overAllContainer = parseInt(window.getComputedStyle(this.componentWrapper).width, 10) -
                    parseInt(window.getComputedStyle(this.componentWrapper).paddingLeft, 10) -
                    parseInt(window.getComputedStyle(this.componentWrapper).paddingRight, 10);
                var remainValue = void 0;
                if (!ej2_base_6.isNullOrUndefined(this.value)) {
                    for (var index = 0; this.value[index]; index++) {
                        data += (index === 0) ? '' : this.delimiterChar + ' ';
                        if (this.mainData && this.mainData.length) {
                            if (this.mode === 'CheckBox') {
                                remainValue = 110;
                                var newTemp = this.listData;
                                this.listData = this.mainData;
                                temp = this.getTextByValue(this.value[index]);
                                this.listData = newTemp;
                            }
                            else {
                                remainValue = 0;
                                temp = this.getTextByValue(this.value[index]);
                            }
                        }
                        else {
                            temp = this.value[index];
                        }
                        data += temp;
                        temp = this.viewWrapper.innerHTML;
                        this.viewWrapper.innerHTML = data;
                        wrapperleng = this.viewWrapper.getBoundingClientRect().width;
                        if ((wrapperleng) > overAllContainer - remainValue) {
                            if (tempData !== undefined) {
                                temp = tempData;
                                index = tempIndex + 1;
                            }
                            this.viewWrapper.innerHTML = temp;
                            remaining = this.value.length - index;
                            break;
                        }
                        else if ((wrapperleng + remainSize) <= overAllContainer) {
                            tempData = data;
                            tempIndex = index;
                        }
                        else if (index === 0) {
                            tempData = '';
                            tempIndex = -1;
                        }
                    }
                }
                if (remaining > 0) {
                    raminElement.innerHTML = '';
                    raminElement.appendChild(compiledString({ 'count': remaining })[0]);
                    this.viewWrapper.appendChild(raminElement);
                }
            }
            else {
                this.viewWrapper.innerHTML = '';
                this.viewWrapper.style.display = 'none';
            }
        };
        MultiSelect.prototype.unWireEvent = function () {
            ej2_base_3.EventHandler.remove(this.componentWrapper, 'mousedown', this.wrapperClick);
            ej2_base_3.EventHandler.remove(window, 'resize', this.windowResize);
            ej2_base_3.EventHandler.remove(this.inputElement, 'focus', this.focusIn);
            ej2_base_3.EventHandler.remove(this.inputElement, 'keydown', this.onKeyDown);
            ej2_base_3.EventHandler.remove(this.inputElement, 'input', this.onInput);
            ej2_base_3.EventHandler.remove(this.inputElement, 'keyup', this.KeyUp);
            ej2_base_3.EventHandler.remove(this.inputElement, 'blur', this.onBlur);
            ej2_base_3.EventHandler.remove(this.componentWrapper, 'mousemove', this.mouseIn);
            ej2_base_3.EventHandler.remove(this.componentWrapper, 'mouseout', this.mouseOut);
            ej2_base_3.EventHandler.remove(this.overAllClear, 'mousedown', this.ClearAll);
        };
        MultiSelect.prototype.selectAllItem = function (state) {
            var li;
            li = this.list.querySelectorAll(state ?
                'li[aria-selected="false"]:not(.e-reorder-hide)' :
                'li[aria-selected="true"]:not(.e-reorder-hide)');
            var length = li.length;
            if (li && li.length) {
                while (length > 0) {
                    this.updateListSelection(li[length - 1], null);
                    length--;
                }
            }
            if (this.mode !== 'Box' && !this.isPopupOpen()) {
                this.updateDelimView();
            }
            else {
                this.searchWrapper.classList.remove(ZERO_SIZE);
            }
            if (this.mode === 'CheckBox') {
                this.updateDelimView();
                this.refreshInputHight();
                this.updateDelimeter(this.delimiterChar);
            }
            else {
                this.updateDelimeter(this.delimiterChar);
            }
            this.refreshPlaceHolder();
        };
        MultiSelect.prototype.setZIndex = function () {
            if (this.popupObj) {
                this.popupObj.setProperties({ 'zIndex': this.zIndex });
            }
        };
        MultiSelect.prototype.updateDataSource = function (prop) {
            this.resetList(this.dataSource);
            if (this.value && this.value.length) {
                this.refreshSelection();
            }
        };
        MultiSelect.prototype.onLoadSelect = function () {
            this.setDynValue = true;
            this.renderPopup();
        };
        MultiSelect.prototype.getPersistData = function () {
            return this.addOnPersist(['value']);
        };
        ;
        MultiSelect.prototype.onPropertyChanged = function (newProp, oldProp) {
            if (newProp.dataSource && !ej2_base_6.isNullOrUndefined(Object.keys(newProp.dataSource))) {
                this.mainList = null;
                this.mainData = null;
            }
            this.setUpdateInitial(['query', 'dataSource'], newProp);
            for (var _i = 0, _a = Object.keys(newProp); _i < _a.length; _i++) {
                var prop = _a[_i];
                switch (prop) {
                    case 'query':
                    case 'dataSource':
                        break;
                    case 'htmlAttributes':
                        this.updateHTMLAttribute();
                        break;
                    case 'showClearButton':
                        this.updateClearButton(newProp.showClearButton);
                        break;
                    case 'value':
                        if (!this.list) {
                            if (this.dataSource instanceof ej2_data_1.DataManager) {
                                this.onLoadSelect();
                            }
                        }
                        else if (!this.inputFocus) {
                            this.initialValueUpdate();
                            if (this.mode !== 'Box') {
                                this.updateDelimView();
                            }
                            this.refreshInputHight();
                            this.refreshPlaceHolder();
                            this.updateValueState(null, this.value, oldProp.value);
                        }
                        break;
                    case 'width':
                        ej2_base_5.setStyleAttribute(this.overAllWrapper, { 'width': ej2_base_6.formatUnit(newProp.width) });
                        this.popupObj.setProperties({ width: this.calcPopupWidth() });
                        break;
                    case 'placeholder':
                        this.refreshPlaceHolder();
                        break;
                    case 'delimiterChar':
                        if (this.mode !== 'Box') {
                            this.updateDelimView();
                        }
                        this.updateData(newProp.delimiterChar);
                        break;
                    case 'cssClass':
                        this.popupWrapper.classList.remove(oldProp.cssClass);
                        this.overAllWrapper.classList.remove(oldProp.cssClass);
                        this.updateCssClass();
                        break;
                    case 'enableRtl':
                        this.enableRTL(newProp.enableRtl);
                        _super.prototype.onPropertyChanged.call(this, newProp, oldProp);
                        break;
                    case 'readonly':
                        this.updateReadonly(newProp.readonly);
                        this.hidePopup();
                        break;
                    case 'enabled':
                        this.hidePopup();
                        this.enable(newProp.enabled);
                        break;
                    case 'showSelectAll':
                        this.popupObj.destroy();
                        this.popupObj = null;
                        this.renderPopup();
                        break;
                    case 'showDropDownIcon':
                        this.dropDownIcon();
                        break;
                    case 'floatLabelType':
                        this.setFloatLabelType();
                        break;
                    case 'enableSelectionOrder':
                        break;
                    default:
                        var msProps = void 0;
                        msProps = this.getPropObject(prop, newProp, oldProp);
                        _super.prototype.onPropertyChanged.call(this, msProps.newProperty, msProps.oldProperty);
                        break;
                }
            }
        };
        MultiSelect.prototype.hidePopup = function () {
            var delay = 100;
            if (this.isPopupOpen()) {
                var eventArgs = { popup: this.popupObj, cancel: false };
                this.trigger('close', eventArgs);
                if (eventArgs.cancel) {
                    return;
                }
                var animModel = {
                    name: 'FadeOut',
                    duration: 100,
                    delay: delay ? delay : 0
                };
                this.beforePopupOpen = false;
                this.overAllWrapper.classList.remove(iconAnimation);
                this.popupObj.hide(new ej2_base_4.Animation(animModel));
                ej2_base_1.attributes(this.inputElement, { 'aria-expanded': 'false' });
                this.notify('inputFocus', { module: 'CheckBoxSelection', enable: this.mode === 'CheckBox', value: 'clear' });
                this.popupObj.hide();
                ej2_base_5.removeClass([document.body, this.popupObj.element], 'e-popup-full-page');
            }
        };
        MultiSelect.prototype.showPopup = function () {
            if (!this.ulElement) {
                this.beforePopupOpen = true;
                _super.prototype.render.call(this);
                return;
            }
            var mainLiLength = this.ulElement.querySelectorAll('li.' + 'e-list-item').length;
            var liLength = this.ulElement.querySelectorAll('li.' + HIDE_LIST).length;
            if (mainLiLength > 0 && (mainLiLength === liLength)) {
                this.beforePopupOpen = false;
                return;
            }
            this.onPopupShown();
        };
        MultiSelect.prototype.selectAll = function (state) {
            var _this = this;
            if (ej2_base_6.isNullOrUndefined(this.list)) {
                this.selectAllAction = function () {
                    if (_this.mode === 'CheckBox' && _this.showSelectAll) {
                        var args = {
                            module: 'CheckBoxSelection',
                            enable: _this.mode === 'CheckBox',
                            value: state ? 'check' : 'uncheck'
                        };
                        _this.notify('checkSelectAll', args);
                    }
                    _this.selectAllItem(state);
                    _this.selectAllAction = null;
                };
                _super.prototype.render.call(this);
            }
            else {
                this.selectAllAction = null;
                if (this.mode === 'CheckBox' && this.showSelectAll) {
                    var args = {
                        module: 'CheckBoxSelection',
                        enable: this.mode === 'CheckBox',
                        value: state ? 'check' : 'uncheck'
                    };
                    this.notify('checkSelectAll', args);
                }
                this.selectAllItem(state);
            }
        };
        MultiSelect.prototype.getModuleName = function () {
            return 'multiselect';
        };
        ;
        MultiSelect.prototype.render = function () {
            var _this = this;
            this.initStatus = false;
            this.setDynValue = false;
            this.searchWrapper = ej2_base_5.createElement('span', { className: SEARCHBOX_WRAPPER });
            this.viewWrapper = ej2_base_5.createElement('span', { className: DELIMITER_VIEW + ' ' + DELIMITER_WRAPPER, styles: 'display:none;' });
            var iconClass;
            this.overAllClear = ej2_base_5.createElement('span', {
                className: CLOSEICON_CLASS, styles: 'display:none;'
            });
            this.componentWrapper = ej2_base_5.createElement('div', { className: ELEMENT_WRAPPER });
            this.overAllWrapper = ej2_base_5.createElement('div', { className: OVER_ALL_WRAPPER });
            if (this.mode === 'CheckBox') {
                ej2_base_5.addClass([this.overAllWrapper], 'e-checkbox');
            }
            if (ej2_base_4.Browser.isDevice) {
                this.componentWrapper.classList.add(ELEMENT_MOBILE_WRAPPER);
            }
            this.overAllWrapper.style.width = ej2_base_6.formatUnit(this.width);
            this.overAllWrapper.appendChild(this.componentWrapper);
            this.popupWrapper = ej2_base_5.createElement('div', { id: this.element.id + '_popup', className: POPUP_WRAPPER });
            if (this.mode === 'Delimiter' || this.mode === 'CheckBox') {
                this.delimiterWrapper = ej2_base_5.createElement('span', { className: DELIMITER_WRAPPER, styles: 'display:none' });
                this.componentWrapper.appendChild(this.delimiterWrapper);
            }
            else {
                this.chipCollectionWrapper = ej2_base_5.createElement('span', {
                    className: CHIP_WRAPPER,
                    styles: 'display:none'
                });
                this.componentWrapper.appendChild(this.chipCollectionWrapper);
            }
            if (this.mode !== 'Box') {
                this.componentWrapper.appendChild(this.viewWrapper);
            }
            this.componentWrapper.appendChild(this.searchWrapper);
            if (this.showClearButton && !ej2_base_4.Browser.isDevice) {
                this.componentWrapper.appendChild(this.overAllClear);
            }
            else {
                this.componentWrapper.classList.add(CLOSE_ICON_HIDE);
            }
            this.dropDownIcon();
            this.inputElement = ej2_base_5.createElement('input', {
                className: INPUT_ELEMENT,
                attrs: {
                    spellcheck: 'false',
                    type: 'text',
                    autocomplete: 'off',
                    tabindex: '0'
                }
            });
            if (this.element.tagName !== this.getNgDirective()) {
                this.element.style.display = 'none';
            }
            if (this.element.tagName === this.getNgDirective()) {
                this.element.appendChild(this.overAllWrapper);
                this.searchWrapper.appendChild(this.inputElement);
            }
            else {
                this.element.parentElement.insertBefore(this.overAllWrapper, this.element);
                this.searchWrapper.appendChild(this.inputElement);
                this.searchWrapper.appendChild(this.element);
                this.element.removeAttribute('tabindex');
            }
            var name = this.element.getAttribute('name') ? this.element.getAttribute('name') : this.element.getAttribute('id');
            var id = this.element.getAttribute('id') ? this.element.getAttribute('id') : ej2_base_6.getUniqueID('ej2_dropdownlist');
            this.element.id = id;
            if (this.floatLabelType !== 'Never') {
                float_label_1.createFloatLabel(this.overAllWrapper, this.searchWrapper, this.element, this.inputElement, this.value, this.floatLabelType, this.placeholder);
            }
            else if (this.floatLabelType === 'Never') {
                this.refreshPlaceHolder();
            }
            this.element.style.opacity = '';
            this.hiddenElement = ej2_base_5.createElement('select', {
                attrs: { 'name': name, 'aria-hidden': 'true', 'class': HIDDEN_ELEMENT, 'tabindex': '-1', 'multiple': 'true' }
            });
            this.element.removeAttribute('name');
            this.componentWrapper.appendChild(this.hiddenElement);
            if (this.mode !== 'CheckBox') {
                this.hideOverAllClear();
            }
            this.wireEvent();
            this.enable(this.enabled);
            this.enableRTL(this.enableRtl);
            if (!(this.dataSource instanceof ej2_data_1.DataManager)) {
                this.renderPopup();
            }
            if (this.value && this.value.length) {
                if (!(this.dataSource instanceof ej2_data_1.DataManager)) {
                    this.initialValueUpdate();
                    this.initialUpdate();
                }
                else {
                    this.setInitialValue = function () {
                        _this.initialValueUpdate();
                        _this.initialUpdate();
                        _this.setInitialValue = null;
                    };
                    _super.prototype.render.call(this);
                }
            }
            else {
                this.initialUpdate();
            }
            this.initStatus = true;
            this.checkAutoFocus();
        };
        ;
        MultiSelect.prototype.checkAutoFocus = function () {
            if (this.element.hasAttribute('autofocus')) {
                this.focusIn();
            }
        };
        MultiSelect.prototype.setFloatLabelType = function () {
            float_label_1.removeFloating(this.overAllWrapper, this.componentWrapper, this.searchWrapper, this.inputElement, this.value, this.floatLabelType, this.placeholder);
            if (this.floatLabelType !== 'Never') {
                float_label_1.createFloatLabel(this.overAllWrapper, this.searchWrapper, this.element, this.inputElement, this.value, this.floatLabelType, this.placeholder);
            }
        };
        MultiSelect.prototype.dropDownIcon = function () {
            if (this.mode === 'CheckBox' && this.showDropDownIcon) {
                this.dropIcon = ej2_base_5.createElement('span', { className: dropdownIcon });
                this.componentWrapper.appendChild(this.dropIcon);
                ej2_base_5.addClass([this.componentWrapper], ['e-down-icon']);
            }
            else {
                if (!ej2_base_6.isNullOrUndefined(this.dropIcon)) {
                    this.dropIcon.parentElement.removeChild(this.dropIcon);
                    ej2_base_5.removeClass([this.componentWrapper], ['e-down-icon']);
                }
            }
        };
        MultiSelect.prototype.initialUpdate = function () {
            if (this.mode !== 'Box') {
                this.updateDelimView();
            }
            this.updateCssClass();
            this.updateHTMLAttribute();
            this.updateReadonly(this.readonly);
            this.refreshInputHight();
        };
        MultiSelect.prototype.destroy = function () {
            if (this.popupObj) {
                this.popupObj.hide();
            }
            this.notify(destroy, {});
            this.unwireListEvents();
            this.unWireEvent();
            this.list = null;
            this.popupObj = null;
            this.mainList = null;
            this.mainData = null;
            _super.prototype.destroy.call(this);
            var temp = ['readonly', 'aria-disabled', 'aria-placeholder', 'placeholder'];
            var length = temp.length;
            while (length > 0) {
                this.inputElement.removeAttribute(temp[length - 1]);
                length--;
            }
            this.element.style.display = 'block';
            if (this.overAllWrapper.parentElement) {
                if (this.overAllWrapper.parentElement.tagName === this.getNgDirective()) {
                    ej2_base_5.remove(this.overAllWrapper);
                }
                else {
                    this.overAllWrapper.parentElement.insertBefore(this.element, this.overAllWrapper);
                    ej2_base_5.remove(this.overAllWrapper);
                }
            }
        };
        ;
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "cssClass", void 0);
        __decorate([
            ej2_base_3.Property('100%')
        ], MultiSelect.prototype, "width", void 0);
        __decorate([
            ej2_base_3.Property('300px')
        ], MultiSelect.prototype, "popupHeight", void 0);
        __decorate([
            ej2_base_3.Property('100%')
        ], MultiSelect.prototype, "popupWidth", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "placeholder", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "filterBarPlaceholder", void 0);
        __decorate([
            ej2_base_3.Property({})
        ], MultiSelect.prototype, "htmlAttributes", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "valueTemplate", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "headerTemplate", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "footerTemplate", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "itemTemplate", void 0);
        __decorate([
            ej2_base_3.Property(false)
        ], MultiSelect.prototype, "allowFiltering", void 0);
        __decorate([
            ej2_base_3.Property(false)
        ], MultiSelect.prototype, "allowCustomValue", void 0);
        __decorate([
            ej2_base_3.Property(true)
        ], MultiSelect.prototype, "showClearButton", void 0);
        __decorate([
            ej2_base_3.Property(1000)
        ], MultiSelect.prototype, "maximumSelectionLength", void 0);
        __decorate([
            ej2_base_3.Property(false)
        ], MultiSelect.prototype, "readonly", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "text", void 0);
        __decorate([
            ej2_base_3.Property(null)
        ], MultiSelect.prototype, "value", void 0);
        __decorate([
            ej2_base_3.Property(true)
        ], MultiSelect.prototype, "hideSelectedItem", void 0);
        __decorate([
            ej2_base_3.Property(true)
        ], MultiSelect.prototype, "closePopupOnSelect", void 0);
        __decorate([
            ej2_base_3.Property('Default')
        ], MultiSelect.prototype, "mode", void 0);
        __decorate([
            ej2_base_3.Property(',')
        ], MultiSelect.prototype, "delimiterChar", void 0);
        __decorate([
            ej2_base_3.Property(true)
        ], MultiSelect.prototype, "ignoreCase", void 0);
        __decorate([
            ej2_base_3.Property(false)
        ], MultiSelect.prototype, "showDropDownIcon", void 0);
        __decorate([
            ej2_base_3.Property('Never')
        ], MultiSelect.prototype, "floatLabelType", void 0);
        __decorate([
            ej2_base_3.Property(false)
        ], MultiSelect.prototype, "showSelectAll", void 0);
        __decorate([
            ej2_base_3.Property('Select All')
        ], MultiSelect.prototype, "selectAllText", void 0);
        __decorate([
            ej2_base_3.Property('Unselect All')
        ], MultiSelect.prototype, "unSelectAllText", void 0);
        __decorate([
            ej2_base_3.Property(true)
        ], MultiSelect.prototype, "enableSelectionOrder", void 0);
        __decorate([
            ej2_base_3.Property(true)
        ], MultiSelect.prototype, "openOnClick", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "change", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "removing", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "removed", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "open", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "close", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "blur", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "focus", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "chipSelection", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "filtering", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "tagging", void 0);
        __decorate([
            ej2_base_3.Event()
        ], MultiSelect.prototype, "customValueSelection", void 0);
        MultiSelect = __decorate([
            ej2_base_2.NotifyPropertyChanges
        ], MultiSelect);
        return MultiSelect;
    }(drop_down_base_1.DropDownBase));
    exports.MultiSelect = MultiSelect;
});
