// @syncfusion/ej2-pdf-export/src/implementation/graphics/fonts/enum.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PdfFontStyle;
    (function (PdfFontStyle) {
        PdfFontStyle[PdfFontStyle["Regular"] = 0] = "Regular";
        PdfFontStyle[PdfFontStyle["Bold"] = 1] = "Bold";
        PdfFontStyle[PdfFontStyle["Italic"] = 2] = "Italic";
        PdfFontStyle[PdfFontStyle["Underline"] = 4] = "Underline";
        PdfFontStyle[PdfFontStyle["Strikeout"] = 8] = "Strikeout";
    })(PdfFontStyle = exports.PdfFontStyle || (exports.PdfFontStyle = {}));
    var PdfFontFamily;
    (function (PdfFontFamily) {
        PdfFontFamily[PdfFontFamily["Helvetica"] = 0] = "Helvetica";
        PdfFontFamily[PdfFontFamily["Courier"] = 1] = "Courier";
        PdfFontFamily[PdfFontFamily["TimesRoman"] = 2] = "TimesRoman";
        PdfFontFamily[PdfFontFamily["Symbol"] = 3] = "Symbol";
        PdfFontFamily[PdfFontFamily["ZapfDingbats"] = 4] = "ZapfDingbats";
    })(PdfFontFamily = exports.PdfFontFamily || (exports.PdfFontFamily = {}));
    var PdfFontType;
    (function (PdfFontType) {
        PdfFontType[PdfFontType["Standard"] = 0] = "Standard";
        PdfFontType[PdfFontType["TrueType"] = 1] = "TrueType";
        PdfFontType[PdfFontType["TrueTypeEmbedded"] = 2] = "TrueTypeEmbedded";
    })(PdfFontType = exports.PdfFontType || (exports.PdfFontType = {}));
    var PdfWordWrapType;
    (function (PdfWordWrapType) {
        PdfWordWrapType[PdfWordWrapType["None"] = 0] = "None";
        PdfWordWrapType[PdfWordWrapType["Word"] = 1] = "Word";
        PdfWordWrapType[PdfWordWrapType["WordOnly"] = 2] = "WordOnly";
        PdfWordWrapType[PdfWordWrapType["Character"] = 3] = "Character";
    })(PdfWordWrapType = exports.PdfWordWrapType || (exports.PdfWordWrapType = {}));
    var PdfSubSuperScript;
    (function (PdfSubSuperScript) {
        PdfSubSuperScript[PdfSubSuperScript["None"] = 0] = "None";
        PdfSubSuperScript[PdfSubSuperScript["SuperScript"] = 1] = "SuperScript";
        PdfSubSuperScript[PdfSubSuperScript["SubScript"] = 2] = "SubScript";
    })(PdfSubSuperScript = exports.PdfSubSuperScript || (exports.PdfSubSuperScript = {}));
    var FontEncoding;
    (function (FontEncoding) {
        FontEncoding[FontEncoding["Unknown"] = 0] = "Unknown";
        FontEncoding[FontEncoding["StandardEncoding"] = 1] = "StandardEncoding";
        FontEncoding[FontEncoding["MacRomanEncoding"] = 2] = "MacRomanEncoding";
        FontEncoding[FontEncoding["MacExpertEncoding"] = 3] = "MacExpertEncoding";
        FontEncoding[FontEncoding["WinAnsiEncoding"] = 4] = "WinAnsiEncoding";
        FontEncoding[FontEncoding["PdfDocEncoding"] = 5] = "PdfDocEncoding";
        FontEncoding[FontEncoding["IdentityH"] = 6] = "IdentityH";
    })(FontEncoding = exports.FontEncoding || (exports.FontEncoding = {}));
});
