// @syncfusion/ej2-ng-inputs/src/color-picker/colorpicker.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, Renderer2, ChangeDetectionStrategy, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ComponentBase, applyMixins, FormBase, setValue } from '@syncfusion/ej2-ng-base';
import { ColorPicker } from '@syncfusion/ej2-inputs';
export var inputs = ['columns', 'cssClass', 'disabled', 'enablePersistence', 'enableRtl', 'inline', 'locale', 'mode', 'modeSwitcher', 'noColor', 'presetColors', 'showButtons', 'value'];
export var outputs = ['focus', 'blur', 'beforeClose', 'beforeModeSwitch', 'beforeOpen', 'beforeTileRender', 'change', 'open', 'select', 'valueChange'];
export var twoWays = ['value'];
/**
 * Represents the EJ2 Angular ColorPicker Component.
 * ```html
 * <input ejs-colorpicker type='color'/>
 * ```
 */
var ColorPickerComponent = /** @class */ (function (_super) {
    __extends(ColorPickerComponent, _super);
    function ColorPickerComponent(ngEle, srenderer, viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.srenderer = srenderer;
        _this.viewContainerRef = viewContainerRef;
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 3; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    ColorPickerComponent.prototype.registerOnChange = function (registerFunction) {
    };
    ColorPickerComponent.prototype.registerOnTouched = function (registerFunction) {
    };
    ColorPickerComponent.prototype.writeValue = function (value) {
    };
    ColorPickerComponent.prototype.setDisabledState = function (disabled) {
    };
    ColorPickerComponent.prototype.ngOnInit = function () {
    };
    ColorPickerComponent.prototype.ngAfterViewInit = function () {
    };
    ColorPickerComponent.prototype.ngOnDestroy = function () {
    };
    ColorPickerComponent.prototype.ngAfterContentChecked = function () {
    };
    ColorPickerComponent.decorators = [
        { type: Component, args: [{
                    selector: '[ejs-colorpicker]',
                    inputs: inputs,
                    outputs: outputs,
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return ColorPickerComponent; }),
                            multi: true
                        }
                    ],
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    ColorPickerComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: ViewContainerRef, },
    ]; };
    return ColorPickerComponent;
}(ColorPicker));
export { ColorPickerComponent };
applyMixins(ColorPickerComponent, [ComponentBase, FormBase]);
