// @syncfusion/ej2-calendars/src/calendar/calendar.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
define(["require", "exports", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base"], function (require, exports, ej2_base_1, ej2_base_2, ej2_base_3, ej2_base_4, ej2_base_5, ej2_base_6, ej2_base_7) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OTHERMONTH = 'e-other-month';
    var ROOT = 'e-calendar';
    var DEVICE = 'e-device';
    var HEADER = 'e-header';
    var RTL = 'e-rtl';
    var CONTENT = 'e-content';
    var YEAR = 'e-year';
    var MONTH = 'e-month';
    var DECADE = 'e-decade';
    var ICON = 'e-icons';
    var PREVICON = 'e-prev';
    var NEXTICON = 'e-next';
    var PREVSPAN = 'e-date-icon-prev';
    var NEXTSPAN = 'e-date-icon-next ';
    var ICONCONTAINER = 'e-icon-container';
    var DISABLED = 'e-disabled';
    var OVERLAY = 'e-overlay';
    var WEEKEND = 'e-weekend';
    var WEEKNUMBER = 'e-week-number';
    var SELECTED = 'e-selected';
    var FOCUSEDDATE = 'e-focused-date';
    var OTHERMONTHROW = 'e-month-hide';
    var TODAY = 'e-today';
    var TITLE = 'e-title';
    var LINK = 'e-day';
    var CELL = 'e-cell';
    var WEEKHEADER = 'e-week-header';
    var ZOOMIN = 'e-zoomin';
    var FOOTER = 'e-footer-container';
    var BTN = 'e-btn';
    var FLAT = 'e-flat';
    var PRIMARY = 'e-primary';
    var dayMilliSeconds = 86400000;
    var minutesMilliSeconds = 60000;
    var CalendarBase = (function (_super) {
        __extends(CalendarBase, _super);
        function CalendarBase(options, element) {
            var _this = _super.call(this, options, element) || this;
            _this.effect = '';
            _this.keyConfigs = {
                controlUp: 'ctrl+38',
                controlDown: 'ctrl+40',
                moveDown: 'downarrow',
                moveUp: 'uparrow',
                moveLeft: 'leftarrow',
                moveRight: 'rightarrow',
                select: 'enter',
                home: 'home',
                end: 'end',
                pageUp: 'pageup',
                pageDown: 'pagedown',
                shiftPageUp: 'shift+pageup',
                shiftPageDown: 'shift+pagedown',
                controlHome: 'ctrl+home',
                controlEnd: 'ctrl+end',
                altUpArrow: 'alt+uparrow',
                spaceBar: 'space'
            };
            return _this;
        }
        CalendarBase.prototype.render = function () {
            this.globalize = new ej2_base_1.Internationalization(this.locale);
            this.todayDisabled = false;
            this.todayDate = new Date(new Date().setHours(0, 0, 0, 0));
            if (this.getModuleName() === 'calendar') {
                this.element.classList.add(ROOT);
                if (this.enableRtl) {
                    this.element.classList.add(RTL);
                }
                if (ej2_base_6.Browser.isDevice) {
                    this.element.classList.add(DEVICE);
                }
                ej2_base_5.attributes(this.element, {
                    'data-role': 'calendar'
                });
            }
            else {
                this.calendarElement = ej2_base_5.createElement('div');
                this.calendarElement.classList.add(ROOT);
                if (this.enableRtl) {
                    this.calendarElement.classList.add(RTL);
                }
                if (ej2_base_6.Browser.isDevice) {
                    this.calendarElement.classList.add(DEVICE);
                }
                ej2_base_5.attributes(this.calendarElement, {
                    'role': 'calendar'
                });
            }
            this.createHeader();
            this.createContent();
            this.wireEvents();
        };
        CalendarBase.prototype.validateDate = function (value) {
            this.setProperties({ min: this.checkDateValue(new Date('' + this.min)) }, true);
            this.setProperties({ max: this.checkDateValue(new Date('' + this.max)) }, true);
            this.currentDate = this.currentDate ? this.currentDate : new Date(new Date().setHours(0, 0, 0, 0));
            if (!ej2_base_7.isNullOrUndefined(value) && this.min <= this.max && value >= this.min && value <= this.max) {
                this.currentDate = new Date('' + value);
            }
        };
        CalendarBase.prototype.minMaxUpdate = function (value) {
            if (!(+this.min <= +this.max)) {
                this.setProperties({ min: this.min }, true);
                ej2_base_5.addClass([this.element], OVERLAY);
            }
            else {
                ej2_base_5.removeClass([this.element], OVERLAY);
            }
            this.min = ej2_base_7.isNullOrUndefined(this.min) || !(+this.min) ? this.min = new Date(1900, 0, 1) : this.min;
            this.max = ej2_base_7.isNullOrUndefined(this.max) || !(+this.max) ? this.max = new Date(2099, 11, 31) : this.max;
            if (+this.min <= +this.max && value && +value <= +this.max && +value >= +this.min) {
                this.currentDate = new Date('' + value);
            }
            else {
                if (+this.min <= +this.max && !value && +this.currentDate > +this.max) {
                    this.currentDate = new Date('' + this.max);
                }
                else {
                    if (+this.currentDate < +this.min) {
                        this.currentDate = new Date('' + this.min);
                    }
                }
            }
        };
        CalendarBase.prototype.createHeader = function () {
            var ariaPrevAttrs = {
                'aria-disabled': 'false',
                'aria-label': 'previous month'
            };
            var ariaNextAttrs = {
                'aria-disabled': 'false',
                'aria-label': 'next month'
            };
            var ariaTitleAttrs = {
                'aria-atomic': 'true', 'aria-live': 'assertive', 'aria-label': 'title'
            };
            this.headerElement = ej2_base_5.createElement('div', { className: HEADER });
            var iconContainer = ej2_base_5.createElement('div', { className: ICONCONTAINER });
            this.previousIcon = ej2_base_5.createElement('button', { className: '' + PREVICON, attrs: { type: 'button' } });
            ej2_base_4.rippleEffect(this.previousIcon, {
                duration: 400,
                selector: '.e-prev',
                isCenterRipple: true
            });
            ej2_base_5.attributes(this.previousIcon, ariaPrevAttrs);
            this.nextIcon = ej2_base_5.createElement('button', { className: '' + NEXTICON, attrs: { type: 'button' } });
            ej2_base_4.rippleEffect(this.nextIcon, {
                selector: '.e-next',
                duration: 400,
                isCenterRipple: true
            });
            ej2_base_5.attributes(this.nextIcon, ariaNextAttrs);
            this.headerTitleElement = ej2_base_5.createElement('div', { className: '' + LINK + ' ' + TITLE });
            ej2_base_5.attributes(this.headerTitleElement, ariaTitleAttrs);
            this.headerElement.appendChild(this.headerTitleElement);
            this.previousIcon.appendChild(ej2_base_5.createElement('span', { className: '' + PREVSPAN + ' ' + ICON }));
            this.nextIcon.appendChild(ej2_base_5.createElement('span', { className: '' + NEXTSPAN + ' ' + ICON }));
            iconContainer.appendChild(this.previousIcon);
            iconContainer.appendChild(this.nextIcon);
            this.headerElement.appendChild(iconContainer);
            if (this.getModuleName() === 'calendar') {
                this.element.appendChild(this.headerElement);
            }
            else {
                this.calendarElement.appendChild(this.headerElement);
            }
        };
        CalendarBase.prototype.createContent = function () {
            this.contentElement = ej2_base_5.createElement('div', { className: CONTENT });
            this.table = ej2_base_5.createElement('table', { attrs: { tabIndex: '0', 'role': 'grid', 'aria-activedescendant': '' } });
            if (this.getModuleName() === 'calendar') {
                this.element.appendChild(this.contentElement);
            }
            else {
                this.calendarElement.appendChild(this.contentElement);
            }
            this.contentElement.appendChild(this.table);
            this.createContentHeader();
            this.createContentBody();
            if (this.showTodayButton) {
                this.createContentFooter();
            }
        };
        CalendarBase.prototype.getCultureValues = function () {
            var culShortNames = [];
            var cldrObj;
            if (this.locale === 'en' || this.locale === 'en-US') {
                cldrObj = (ej2_base_6.getValue('days.stand-alone.short', ej2_base_4.getDefaultDateObject()));
            }
            else {
                cldrObj = (this.getCultureObjects(ej2_base_4.cldrData, '' + this.locale));
            }
            for (var _i = 0, _a = Object.keys(cldrObj); _i < _a.length; _i++) {
                var obj = _a[_i];
                culShortNames.push(ej2_base_6.getValue(obj, cldrObj));
            }
            return culShortNames;
        };
        CalendarBase.prototype.createContentHeader = function () {
            if (this.getModuleName() === 'calendar') {
                if (!ej2_base_7.isNullOrUndefined(this.element.querySelectorAll('.e-content .e-week-header')[0])) {
                    ej2_base_5.detach(this.element.querySelectorAll('.e-content .e-week-header')[0]);
                }
            }
            else {
                if (!ej2_base_7.isNullOrUndefined(this.calendarElement.querySelectorAll('.e-content .e-week-header')[0])) {
                    ej2_base_5.detach(this.calendarElement.querySelectorAll('.e-content .e-week-header')[0]);
                }
            }
            var daysCount = 6;
            var html = '';
            var shortNames;
            if (this.firstDayOfWeek > 6 || this.firstDayOfWeek < 0) {
                this.setProperties({ firstDayOfWeek: 0 }, true);
            }
            this.tableHeadElement = ej2_base_5.createElement('thead', { className: WEEKHEADER });
            if (this.weekNumber) {
                html += '<th class="e-week-number"></th>';
                if (this.getModuleName() === 'calendar') {
                    ej2_base_5.addClass([this.element], '' + WEEKNUMBER);
                }
                else {
                    ej2_base_5.addClass([this.calendarElement], '' + WEEKNUMBER);
                }
            }
            shortNames = this.shiftArray(((this.getCultureValues().length > 0 && this.getCultureValues())), this.firstDayOfWeek);
            for (var days = 0; days <= daysCount; days++) {
                html += '<th  class="">' + shortNames[days] + '</th>';
            }
            html = '<tr>' + html + '</tr>';
            this.tableHeadElement.innerHTML = html;
            this.table.appendChild(this.tableHeadElement);
        };
        CalendarBase.prototype.createContentBody = function () {
            if (this.getModuleName() === 'calendar') {
                if (!ej2_base_7.isNullOrUndefined(this.element.querySelectorAll('.e-content tbody')[0])) {
                    ej2_base_5.detach(this.element.querySelectorAll('.e-content tbody')[0]);
                }
            }
            else {
                if (!ej2_base_7.isNullOrUndefined(this.calendarElement.querySelectorAll('.e-content tbody')[0])) {
                    ej2_base_5.detach(this.calendarElement.querySelectorAll('.e-content tbody')[0]);
                }
            }
            switch (this.start) {
                case 'Year':
                    this.renderYears();
                    break;
                case 'Decade':
                    this.renderDecades();
                    break;
                default:
                    this.renderMonths();
            }
        };
        CalendarBase.prototype.updateFooter = function () {
            this.todayElement.textContent = this.l10.getConstant('today');
        };
        CalendarBase.prototype.createContentFooter = function () {
            if (this.showTodayButton) {
                var minimum = new Date(+this.min);
                var maximum = new Date(+this.max);
                var l10nLocale = { today: 'Today' };
                this.globalize = new ej2_base_1.Internationalization(this.locale);
                this.l10 = new ej2_base_2.L10n(this.getModuleName(), l10nLocale, this.locale);
                this.todayElement = ej2_base_5.createElement('button');
                ej2_base_4.rippleEffect(this.todayElement);
                this.updateFooter();
                ej2_base_5.addClass([this.todayElement], [BTN, TODAY, FLAT, PRIMARY]);
                if ((!(+new Date(minimum.setHours(0, 0, 0, 0)) <= +this.todayDate &&
                    +this.todayDate <= +new Date(maximum.setHours(0, 0, 0, 0)))) || (this.todayDisabled)) {
                    ej2_base_5.addClass([this.todayElement], DISABLED);
                }
                this.footer = ej2_base_5.createElement('div', { className: FOOTER });
                this.footer.appendChild(this.todayElement);
                if (this.getModuleName() === 'calendar') {
                    this.element.appendChild(this.footer);
                }
                if (this.getModuleName() === 'datepicker') {
                    this.calendarElement.appendChild(this.footer);
                }
                if (this.getModuleName() === 'datetimepicker') {
                    this.calendarElement.appendChild(this.footer);
                }
                if (!this.todayElement.classList.contains(DISABLED)) {
                    ej2_base_1.EventHandler.add(this.todayElement, 'click', this.todayButtonClick, this);
                }
            }
        };
        CalendarBase.prototype.wireEvents = function () {
            ej2_base_1.EventHandler.add(this.headerTitleElement, 'click', this.navigateTitle, this);
            if (this.getModuleName() === 'calendar') {
                this.keyboardModule = new ej2_base_2.KeyboardEvents(this.element, {
                    eventName: 'keydown',
                    keyAction: this.keyActionHandle.bind(this),
                    keyConfigs: this.keyConfigs
                });
            }
            else {
                this.keyboardModule = new ej2_base_2.KeyboardEvents(this.calendarElement, {
                    eventName: 'keydown',
                    keyAction: this.keyActionHandle.bind(this),
                    keyConfigs: this.keyConfigs
                });
            }
        };
        CalendarBase.prototype.todayButtonClick = function (value) {
            if (this.showTodayButton) {
                if (this.getViewNumber(this.start) >= this.getViewNumber(this.depth)) {
                    this.navigateTo(this.depth, new Date('' + value));
                }
                else {
                    this.navigateTo('Month', new Date('' + value));
                }
                this.effect = '';
            }
        };
        CalendarBase.prototype.keyActionHandle = function (e, value) {
            var view = this.getViewNumber(this.currentView());
            var focusedDate = this.tableBodyElement.querySelector('tr td.e-focused-date');
            var selectedDate = this.tableBodyElement.querySelector('tr td.e-selected');
            var depthValue = this.getViewNumber(this.depth);
            var levelRestrict = (view === depthValue && this.getViewNumber(this.start) >= depthValue);
            this.effect = '';
            switch (e.action) {
                case 'moveLeft':
                    this.KeyboardNavigate(-1, view, e, this.max, this.min);
                    e.preventDefault();
                    break;
                case 'moveRight':
                    this.KeyboardNavigate(1, view, e, this.max, this.min);
                    e.preventDefault();
                    break;
                case 'moveUp':
                    if (view === 0) {
                        this.KeyboardNavigate(-7, view, e, this.max, this.min);
                    }
                    else {
                        this.KeyboardNavigate(-4, view, e, this.max, this.min);
                    }
                    e.preventDefault();
                    break;
                case 'moveDown':
                    if (view === 0) {
                        this.KeyboardNavigate(7, view, e, this.max, this.min);
                    }
                    else {
                        this.KeyboardNavigate(4, view, e, this.max, this.min);
                    }
                    e.preventDefault();
                    break;
                case 'select':
                    if (e.target === this.todayElement) {
                        this.todayButtonClick(value);
                    }
                    else {
                        var element = !ej2_base_7.isNullOrUndefined(focusedDate) ? focusedDate : selectedDate;
                        if (!ej2_base_7.isNullOrUndefined(element) && !element.classList.contains(DISABLED)) {
                            if (levelRestrict) {
                                var d = new Date(parseInt('' + (element).id, 0));
                                this.selectDate(e, d, (element), value);
                            }
                            else {
                                this.contentClick(null, --view, (element), value);
                            }
                        }
                    }
                    break;
                case 'controlUp':
                    this.title();
                    e.preventDefault();
                    break;
                case 'controlDown':
                    if (!ej2_base_7.isNullOrUndefined(focusedDate) || !ej2_base_7.isNullOrUndefined(selectedDate) && !levelRestrict) {
                        this.contentClick(null, --view, (focusedDate || selectedDate), value);
                    }
                    e.preventDefault();
                    break;
                case 'home':
                    this.currentDate = this.firstDay(this.currentDate);
                    ej2_base_5.detach(this.tableBodyElement);
                    this.renderMonths(e);
                    e.preventDefault();
                    break;
                case 'end':
                    this.currentDate = this.lastDay(this.currentDate);
                    ej2_base_5.detach(this.tableBodyElement);
                    this.renderMonths(e);
                    e.preventDefault();
                    break;
                case 'pageUp':
                    this.addMonths(this.currentDate, -1);
                    this.navigateTo('Month', this.currentDate);
                    e.preventDefault();
                    break;
                case 'pageDown':
                    this.addMonths(this.currentDate, 1);
                    this.navigateTo('Month', this.currentDate);
                    e.preventDefault();
                    break;
                case 'shiftPageUp':
                    this.addYears(this.currentDate, -1);
                    this.navigateTo('Month', this.currentDate);
                    e.preventDefault();
                    break;
                case 'shiftPageDown':
                    this.addYears(this.currentDate, 1);
                    this.navigateTo('Month', this.currentDate);
                    e.preventDefault();
                    break;
                case 'controlHome':
                    this.navigateTo('Month', new Date(this.currentDate.getFullYear(), 0, 1));
                    e.preventDefault();
                    break;
                case 'controlEnd':
                    this.navigateTo('Month', new Date(this.currentDate.getFullYear(), 11, 31));
                    e.preventDefault();
                    break;
            }
            if (this.getModuleName() === 'calendar') {
                this.table.focus();
            }
        };
        CalendarBase.prototype.KeyboardNavigate = function (number, currentView, e, max, min) {
            var date = new Date('' + this.currentDate);
            switch (currentView) {
                case 2:
                    this.addYears(this.currentDate, number);
                    if (this.isMinMaxRange(this.currentDate)) {
                        ej2_base_5.detach(this.tableBodyElement);
                        this.renderDecades(e);
                    }
                    else {
                        this.currentDate = date;
                    }
                    break;
                case 1:
                    this.addMonths(this.currentDate, number);
                    if (this.isMinMaxRange(this.currentDate)) {
                        ej2_base_5.detach(this.tableBodyElement);
                        this.renderYears(e);
                    }
                    else {
                        this.currentDate = date;
                    }
                    break;
                case 0:
                    this.addDay(this.currentDate, number, e, max, min);
                    if (this.isMinMaxRange(this.currentDate)) {
                        ej2_base_5.detach(this.tableBodyElement);
                        this.renderMonths(e);
                    }
                    else {
                        this.currentDate = date;
                    }
                    break;
            }
        };
        CalendarBase.prototype.preRender = function (value) {
            var _this = this;
            this.navigatePreviousHandler = this.navigatePrevious.bind(this);
            this.navigateNextHandler = this.navigateNext.bind(this);
            this.navigateHandler = function (e) {
                _this.triggerNavigate(e);
            };
        };
        ;
        CalendarBase.prototype.minMaxDate = function (localDate) {
            var currentDate = new Date(new Date(+localDate).setHours(0, 0, 0, 0));
            var minDate = new Date(new Date(+this.min).setHours(0, 0, 0, 0));
            var maxDate = new Date(new Date(+this.max).setHours(0, 0, 0, 0));
            if (+currentDate === +minDate || +currentDate === +maxDate) {
                if (+localDate < +this.min) {
                    localDate = new Date(+this.min);
                }
                if (+localDate > +this.max) {
                    localDate = new Date(+this.max);
                }
            }
            return localDate;
        };
        CalendarBase.prototype.renderMonths = function (e, value) {
            var numCells = this.weekNumber ? 8 : 7;
            var tdEles = this.renderDays(this.currentDate, e, value);
            this.createContentHeader();
            this.renderTemplate(tdEles, numCells, MONTH, e, value);
        };
        CalendarBase.prototype.renderDays = function (currentDate, e, value) {
            var tdEles = [];
            var cellsCount = 42;
            var localDate = new Date('' + currentDate);
            var minMaxDate;
            var numCells = this.weekNumber ? 8 : 7;
            var currentMonth = localDate.getMonth();
            this.titleUpdate(currentDate, 'days');
            var d = localDate;
            localDate = new Date(d.getFullYear(), d.getMonth(), 0, d.getHours(), d.getMinutes(), d.getSeconds(), d.getMilliseconds());
            while (localDate.getDay() !== this.firstDayOfWeek) {
                this.setStartDate(localDate, -1 * dayMilliSeconds);
            }
            for (var day = 0; day < cellsCount; ++day) {
                var weekEle = ej2_base_5.createElement('td', { className: CELL });
                var weekAnchor = ej2_base_5.createElement('span');
                if (day % 7 === 0 && this.weekNumber) {
                    weekAnchor.textContent = '' + this.getWeek(localDate);
                    weekEle.appendChild(weekAnchor);
                    ej2_base_5.addClass([weekEle], '' + WEEKNUMBER);
                    tdEles.push(weekEle);
                }
                minMaxDate = new Date(+localDate);
                localDate = this.minMaxDate(localDate);
                var dateFormatOptions = { type: 'dateTime', skeleton: 'full' };
                var date = this.globalize.parseDate(this.globalize.formatDate(localDate, dateFormatOptions), dateFormatOptions);
                var tdEle = this.dayCell(localDate);
                var title = this.globalize.formatDate(localDate, { type: 'date', skeleton: 'full' });
                var dayLink = ej2_base_5.createElement('span');
                dayLink.textContent = this.globalize.formatDate(localDate, { type: 'date', skeleton: 'd' });
                var disabled = (this.min > localDate) || (this.max < localDate);
                if (disabled) {
                    ej2_base_5.addClass([tdEle], DISABLED);
                    ej2_base_5.addClass([tdEle], OVERLAY);
                }
                else {
                    dayLink.setAttribute('title', '' + title);
                }
                if (currentMonth !== localDate.getMonth()) {
                    ej2_base_5.addClass([tdEle], OTHERMONTH);
                }
                if (localDate.getDay() === 0 || localDate.getDay() === 6) {
                    ej2_base_5.addClass([tdEle], WEEKEND);
                }
                tdEle.appendChild(dayLink);
                this.renderDayCellArgs = {
                    date: localDate,
                    isDisabled: false,
                    element: tdEle,
                    isOutOfRange: disabled
                };
                var argument = this.renderDayCellArgs;
                this.renderDayCellEvent(argument);
                if (argument.isDisabled) {
                    if (value && +value === +argument.date) {
                        this.setProperties({ value: null }, true);
                    }
                }
                if (this.renderDayCellArgs.isDisabled && !tdEle.classList.contains(SELECTED)) {
                    ej2_base_5.addClass([tdEle], DISABLED);
                    ej2_base_5.addClass([tdEle], OVERLAY);
                    if (+this.renderDayCellArgs.date === +this.todayDate) {
                        this.todayDisabled = true;
                    }
                }
                var otherMnthBool = tdEle.classList.contains(OTHERMONTH);
                var disabledCls = tdEle.classList.contains(DISABLED);
                if (!disabledCls) {
                    ej2_base_1.EventHandler.add(tdEle, 'click', this.clickHandler, this);
                }
                if (!otherMnthBool && !disabledCls && this.getDateVal(localDate, value)) {
                    ej2_base_5.addClass([tdEle], SELECTED);
                }
                else {
                    if (currentDate.getDate() === localDate.getDate() && !otherMnthBool && !disabledCls) {
                        ej2_base_5.addClass([tdEle], FOCUSEDDATE);
                    }
                    else {
                        if (currentDate >= this.max && parseInt(tdEle.id, 0) === +this.max && !otherMnthBool && !disabledCls) {
                            ej2_base_5.addClass([tdEle], FOCUSEDDATE);
                        }
                        if (currentDate <= this.min && parseInt(tdEle.id, 0) === +this.min && !otherMnthBool && !disabledCls) {
                            ej2_base_5.addClass([tdEle], FOCUSEDDATE);
                        }
                    }
                }
                if (date.getMonth() === new Date().getMonth() && date.getDate() === new Date().getDate()) {
                    if (date.getFullYear() === new Date().getFullYear()) {
                        ej2_base_5.addClass([tdEle], TODAY);
                    }
                }
                tdEles.push(this.renderDayCellArgs.element);
                localDate = new Date(+minMaxDate);
                this.addDay(localDate, 1, null, this.max, this.min);
            }
            return tdEles;
        };
        CalendarBase.prototype.renderYears = function (e, value) {
            this.removeTableHeadElement();
            var numCells = 4;
            var days;
            var tdEles = [];
            var valueUtil = ej2_base_7.isNullOrUndefined(value);
            var curDate = new Date('' + this.currentDate);
            var mon = curDate.getMonth();
            var yr = curDate.getFullYear();
            var localDate = curDate;
            var curYrs = localDate.getFullYear();
            var minYr = new Date('' + this.min).getFullYear();
            var minMonth = new Date('' + this.min).getMonth();
            var maxYr = new Date('' + this.max).getFullYear();
            var maxMonth = new Date('' + this.max).getMonth();
            localDate.setMonth(0);
            this.titleUpdate(this.currentDate, 'months');
            var disabled = (this.min > localDate) || (this.max < localDate);
            localDate.setDate(1);
            for (var month = 0; month < 12; ++month) {
                var tdEle = this.dayCell(localDate);
                var dayLink = ej2_base_5.createElement('span');
                var localMonth = (value && (value).getMonth() === localDate.getMonth());
                var select = (value && (value).getFullYear() === yr && localMonth);
                dayLink.textContent = this.globalize.formatDate(localDate, { type: 'dateTime', skeleton: 'MMM' });
                if ((this.min && (curYrs < minYr || (month < minMonth && curYrs === minYr))) || (this.max && (curYrs > maxYr || (month > maxMonth && curYrs >= maxYr)))) {
                    ej2_base_5.addClass([tdEle], DISABLED);
                }
                else if (!valueUtil && select) {
                    ej2_base_5.addClass([tdEle], SELECTED);
                }
                else {
                    if (localDate.getMonth() === mon && this.currentDate.getMonth() === mon) {
                        ej2_base_5.addClass([tdEle], FOCUSEDDATE);
                    }
                }
                localDate.setDate(1);
                localDate.setMonth(localDate.getMonth() + 1);
                if (!tdEle.classList.contains(DISABLED)) {
                    ej2_base_1.EventHandler.add(tdEle, 'click', this.clickHandler, this);
                }
                tdEle.appendChild(dayLink);
                tdEles.push(tdEle);
            }
            this.renderTemplate(tdEles, numCells, YEAR, e, value);
        };
        CalendarBase.prototype.renderDecades = function (e, value) {
            this.removeTableHeadElement();
            var numCells = 4;
            var yearCell = 12;
            var tdEles = [];
            var localDate = new Date('' + this.currentDate);
            localDate.setMonth(0);
            localDate.setDate(1);
            var localYr = localDate.getFullYear();
            var startYr = new Date('' + (localYr - localYr % 10));
            var endYr = new Date('' + (localYr - localYr % 10 + (10 - 1)));
            var startHdrYr = this.globalize.formatDate(startYr, { type: 'dateTime', skeleton: 'y' });
            var endHdrYr = this.globalize.formatDate(endYr, { type: 'dateTime', skeleton: 'y' });
            this.headerTitleElement.textContent = startHdrYr + ' - ' + (endHdrYr);
            var start = new Date(localYr - (localYr % 10) - 1, 0, 1);
            var startYear = start.getFullYear();
            for (var rowIterator = 0; rowIterator < yearCell; ++rowIterator) {
                var year = startYear + rowIterator;
                localDate.setFullYear(year);
                var tdEle = this.dayCell(localDate);
                ej2_base_5.attributes(tdEle, { 'role': 'gridcell' });
                var dayLink = ej2_base_5.createElement('span');
                dayLink.textContent = this.globalize.formatDate(localDate, { type: 'dateTime', skeleton: 'y' });
                if (year < new Date('' + this.min).getFullYear() || year > new Date('' + this.max).getFullYear()) {
                    ej2_base_5.addClass([tdEle], DISABLED);
                }
                else if (!ej2_base_7.isNullOrUndefined(value) && localDate.getFullYear() === (value).getFullYear()) {
                    ej2_base_5.addClass([tdEle], SELECTED);
                }
                else {
                    if (localDate.getFullYear() === this.currentDate.getFullYear() && !tdEle.classList.contains(DISABLED)) {
                        ej2_base_5.addClass([tdEle], FOCUSEDDATE);
                    }
                }
                if (!tdEle.classList.contains(DISABLED)) {
                    ej2_base_1.EventHandler.add(tdEle, 'click', this.clickHandler, this);
                }
                tdEle.appendChild(dayLink);
                tdEles.push(tdEle);
            }
            this.renderTemplate(tdEles, numCells, 'e-decade', e, value);
        };
        CalendarBase.prototype.dayCell = function (localDate) {
            var dateFormatOptions = { skeleton: 'full', type: 'dateTime' };
            var date = this.globalize.parseDate(this.globalize.formatDate(localDate, dateFormatOptions), dateFormatOptions);
            var value = date.valueOf();
            var attrs = {
                className: CELL, attrs: { 'id': '' + ej2_base_6.getUniqueID('' + value), 'aria-selected': 'false', 'role': 'gridcell' }
            };
            return ej2_base_5.createElement('td', attrs);
        };
        CalendarBase.prototype.firstDay = function (date) {
            var collection = this.tableBodyElement.querySelectorAll('td' + ':not(.' + OTHERMONTH + '');
            if (collection.length) {
                for (var i = 0; i < collection.length; i++) {
                    if (!collection[i].classList.contains(DISABLED)) {
                        date = new Date(parseInt(collection[i].id, 0));
                        break;
                    }
                }
            }
            return date;
        };
        CalendarBase.prototype.lastDay = function (date) {
            var lastDate = new Date(date.getFullYear(), date.getMonth() + 1, 0);
            var timeOffset = Math.abs(lastDate.getTimezoneOffset() - this.firstDay(date).getTimezoneOffset());
            if (timeOffset) {
                lastDate.setHours(this.firstDay(date).getHours() + (timeOffset / 60));
            }
            return this.findLastDay(lastDate);
        };
        ;
        CalendarBase.prototype.checkDateValue = function (value) {
            return (!ej2_base_7.isNullOrUndefined(value) && value instanceof Date && !isNaN(+value)) ? value : null;
        };
        CalendarBase.prototype.findLastDay = function (date) {
            var collection = this.tableBodyElement.querySelectorAll('td' + ':not(.' + OTHERMONTH + '');
            if (collection.length) {
                for (var i = collection.length - 1; i >= 0; i--) {
                    if (!collection[i].classList.contains(DISABLED)) {
                        date = new Date(parseInt(collection[i].id, 0));
                        break;
                    }
                }
            }
            return date;
        };
        CalendarBase.prototype.removeTableHeadElement = function () {
            if (this.getModuleName() === 'calendar') {
                if (!ej2_base_7.isNullOrUndefined(this.element.querySelectorAll('.e-content table thead')[0])) {
                    ej2_base_5.detach(this.tableHeadElement);
                }
            }
            else {
                if (!ej2_base_7.isNullOrUndefined(this.calendarElement.querySelectorAll('.e-content table thead')[0])) {
                    ej2_base_5.detach(this.tableHeadElement);
                }
            }
        };
        CalendarBase.prototype.renderTemplate = function (elements, count, classNm, e, value) {
            var view = this.getViewNumber(this.currentView());
            var trEle;
            this.tableBodyElement = ej2_base_5.createElement('tbody');
            this.table.appendChild(this.tableBodyElement);
            ej2_base_5.removeClass([this.contentElement, this.headerElement], [MONTH, DECADE, YEAR]);
            ej2_base_5.addClass([this.contentElement, this.headerElement], [classNm]);
            var weekNumCell = 41;
            var numberCell = 35;
            var otherMonthCell = 6;
            var row = count;
            var rowIterator = 0;
            for (var dayCell = 0; dayCell < elements.length / count; ++dayCell) {
                trEle = ej2_base_5.createElement('tr', { attrs: { 'role': 'row' } });
                for (rowIterator = 0 + rowIterator; rowIterator < row; rowIterator++) {
                    if (!elements[rowIterator].classList.contains('e-week-number') && !ej2_base_7.isNullOrUndefined(elements[rowIterator].children[0])) {
                        ej2_base_5.addClass([elements[rowIterator].children[0]], [LINK]);
                        ej2_base_4.rippleEffect(elements[rowIterator].children[0], {
                            duration: 600,
                            isCenterRipple: true
                        });
                    }
                    trEle.appendChild(elements[rowIterator]);
                    if (!this.weekNumber && rowIterator === otherMonthCell && elements[otherMonthCell].classList.contains(OTHERMONTH)) {
                        ej2_base_5.addClass([trEle], OTHERMONTHROW);
                    }
                    if (this.weekNumber) {
                        if (rowIterator === weekNumCell && elements[weekNumCell].classList.contains(OTHERMONTH)) {
                            ej2_base_5.addClass([trEle], OTHERMONTHROW);
                        }
                    }
                    else {
                        if (rowIterator === numberCell && elements[numberCell].classList.contains(OTHERMONTH)) {
                            ej2_base_5.addClass([trEle], OTHERMONTHROW);
                        }
                    }
                }
                row = row + count;
                rowIterator = rowIterator + 0;
                this.tableBodyElement.appendChild(trEle);
            }
            this.table.querySelector('tbody').className = this.effect;
            this.iconHandler();
            if (view !== this.getViewNumber(this.currentView()) || (view === 0 && view !== this.getViewNumber(this.currentView()))) {
                this.navigateHandler(e);
            }
            this.setAriaActiveDescendant();
        };
        CalendarBase.prototype.clickHandler = function (e, value) {
            this.clickEventEmitter(e);
            var eve = e.currentTarget;
            var view = this.getViewNumber(this.currentView());
            if (eve.classList.contains(OTHERMONTH)) {
                this.contentClick(e, 0, null, value);
            }
            else if (view === this.getViewNumber(this.depth) && this.getViewNumber(this.start) >= this.getViewNumber(this.depth)) {
                this.contentClick(e, 1, null, value);
            }
            else if (2 === view) {
                this.contentClick(e, 1, null, value);
            }
            else if (!eve.classList.contains(OTHERMONTH) && view === 0) {
                this.selectDate(e, this.getIdValue(e, null), null, value);
            }
            else {
                this.contentClick(e, 0, eve, value);
            }
            if (this.getModuleName() === 'calendar') {
                this.table.focus();
            }
        };
        CalendarBase.prototype.clickEventEmitter = function (e) {
            e.preventDefault();
        };
        CalendarBase.prototype.contentClick = function (e, view, element, value) {
            var currentView = this.getViewNumber(this.currentView());
            var d = this.getIdValue(e, element);
            switch (view) {
                case 0:
                    if (currentView === this.getViewNumber(this.depth) && this.getViewNumber(this.start) >= this.getViewNumber(this.depth)) {
                        ej2_base_5.detach(this.tableBodyElement);
                        this.currentDate = d;
                        this.effect = ZOOMIN;
                        this.renderMonths(e);
                    }
                    else {
                        this.currentDate.setMonth(d.getMonth());
                        if (d.getMonth() > 0 && this.currentDate.getMonth() !== d.getMonth()) {
                            this.currentDate.setDate(0);
                        }
                        this.currentDate.setFullYear(d.getFullYear());
                        this.effect = ZOOMIN;
                        ej2_base_5.detach(this.tableBodyElement);
                        this.renderMonths(e);
                    }
                    break;
                case 1:
                    if (currentView === this.getViewNumber(this.depth) && this.getViewNumber(this.start) >= this.getViewNumber(this.depth)) {
                        this.selectDate(e, d, null, value);
                    }
                    else {
                        this.currentDate.setFullYear(d.getFullYear());
                        this.effect = ZOOMIN;
                        ej2_base_5.detach(this.tableBodyElement);
                        this.renderYears(e);
                    }
            }
        };
        CalendarBase.prototype.switchView = function (view, e) {
            switch (view) {
                case 0:
                    ej2_base_5.detach(this.tableBodyElement);
                    this.renderMonths(e);
                    break;
                case 1:
                    ej2_base_5.detach(this.tableBodyElement);
                    this.renderYears(e);
                    break;
                case 2:
                    ej2_base_5.detach(this.tableBodyElement);
                    this.renderDecades(e);
            }
        };
        CalendarBase.prototype.getModuleName = function () {
            return 'calendar';
        };
        CalendarBase.prototype.getPersistData = function () {
            var keyEntity = ['value'];
            return this.addOnPersist(keyEntity);
        };
        CalendarBase.prototype.onPropertyChanged = function (newProp, oldProp) {
            this.effect = '';
            for (var _i = 0, _a = Object.keys(newProp); _i < _a.length; _i++) {
                var prop = _a[_i];
                switch (prop) {
                    case 'enableRtl':
                        if (newProp.enableRtl) {
                            if (this.getModuleName() === 'calendar') {
                                this.element.classList.add('e-rtl');
                            }
                            else {
                                this.calendarElement.classList.add('e-rtl');
                            }
                        }
                        else {
                            if (this.getModuleName() === 'calendar') {
                                this.element.classList.remove('e-rtl');
                            }
                            else {
                                this.calendarElement.classList.remove('e-rtl');
                            }
                        }
                        break;
                    case 'min':
                    case 'max':
                        prop === 'min' ? this.setProperties({ min: this.checkDateValue(new Date('' + newProp.min)) }, true) :
                            this.setProperties({ max: this.checkDateValue(new Date('' + newProp.max)) }, true);
                        this.setProperties({ start: this.currentView() }, true);
                        ej2_base_5.detach(this.tableBodyElement);
                        this.minMaxUpdate();
                        this.createContentBody();
                        if ((this.todayDate < this.min || this.max < this.todayDate) && (this.footer) && (this.todayElement)) {
                            ej2_base_5.detach(this.todayElement);
                            ej2_base_5.detach(this.footer);
                            this.todayElement = this.footer = undefined;
                            this.createContentFooter();
                        }
                        else {
                            if (this.todayElement.classList.contains('e-disabled') && (this.footer) && (this.todayElement)) {
                                ej2_base_5.removeClass([this.todayElement], DISABLED);
                            }
                        }
                        break;
                    case 'start':
                    case 'weekNumber':
                    case 'firstDayOfWeek':
                        this.createContentHeader();
                        this.createContentBody();
                        break;
                    case 'locale':
                        this.globalize = new ej2_base_1.Internationalization(this.locale);
                        this.createContentHeader();
                        this.createContentBody();
                        this.l10.setLocale(this.locale);
                        this.updateFooter();
                        break;
                    case 'showTodayButton':
                        if (newProp.showTodayButton) {
                            this.createContentFooter();
                        }
                        else {
                            ej2_base_5.detach(this.todayElement);
                            ej2_base_5.detach(this.footer);
                            this.todayElement = this.footer = undefined;
                        }
                        this.setProperties({ showTodayButton: newProp.showTodayButton }, true);
                        break;
                }
            }
        };
        CalendarBase.prototype.setValueUpdate = function () {
            ej2_base_5.detach(this.tableBodyElement);
            this.setProperties({ start: this.currentView() }, true);
            this.createContentBody();
        };
        CalendarBase.prototype.titleUpdate = function (date, view) {
            var globalize = new ej2_base_1.Internationalization(this.locale);
            switch (view) {
                case 'days':
                    this.headerTitleElement.textContent = globalize.formatDate(date, { type: 'dateTime', skeleton: 'yMMMM' });
                    break;
                case 'months':
                    this.headerTitleElement.textContent = globalize.formatDate(date, { type: 'dateTime', skeleton: 'y' });
            }
        };
        CalendarBase.prototype.setActiveDescendant = function () {
            var id;
            var focusedEle = this.tableBodyElement.querySelector('tr td.e-focused-date');
            var selectedEle = this.tableBodyElement.querySelector('tr td.e-selected');
            var title = this.globalize.formatDate(this.currentDate, { type: 'date', skeleton: 'full' });
            if (selectedEle || focusedEle) {
                (focusedEle || selectedEle).setAttribute('aria-selected', 'true');
                (focusedEle || selectedEle).setAttribute('aria-label', 'The current focused date is ' + '' + title);
                id = (focusedEle || selectedEle).getAttribute('id');
            }
            return id;
        };
        CalendarBase.prototype.iconHandler = function () {
            new Date('' + this.currentDate).setDate(1);
            switch (this.currentView()) {
                case 'Month':
                    this.previousIconHandler(this.compareMonth(new Date('' + this.currentDate), this.min) < 1);
                    this.nextIconHandler(this.compareMonth(new Date('' + this.currentDate), this.max) > -1);
                    break;
                case 'Year':
                    this.previousIconHandler(this.compareYear(new Date('' + this.currentDate), this.min) < 1);
                    this.nextIconHandler(this.compareYear(new Date('' + this.currentDate), this.max) > -1);
                    break;
                case 'Decade':
                    this.previousIconHandler(this.compareDecade(new Date('' + this.currentDate), this.min) < 1);
                    this.nextIconHandler(this.compareDecade(new Date('' + this.currentDate), this.max) > -1);
            }
        };
        CalendarBase.prototype.destroy = function () {
            if (this.getModuleName() === 'calendar') {
                ej2_base_5.removeClass([this.element], [ROOT]);
            }
            else {
                if (this.calendarElement) {
                    ej2_base_5.removeClass([this.element], [ROOT]);
                }
            }
            if (this.getModuleName() === 'calendar') {
                ej2_base_1.EventHandler.remove(this.headerTitleElement, 'click', this.navigateTitle);
                if (this.todayElement) {
                    ej2_base_1.EventHandler.remove(this.todayElement, 'click', this.todayButtonClick);
                }
                this.previousIconHandler(true);
                this.nextIconHandler(true);
                this.keyboardModule.destroy();
                this.element.removeAttribute('data-role');
            }
            this.element.innerHTML = '';
            _super.prototype.destroy.call(this);
        };
        CalendarBase.prototype.title = function (e) {
            var currentView = this.getViewNumber(this.currentView());
            this.effect = ZOOMIN;
            this.switchView(++currentView, e);
        };
        CalendarBase.prototype.getViewNumber = function (stringVal) {
            if (stringVal === 'Month') {
                return 0;
            }
            else if (stringVal === 'Year') {
                return 1;
            }
            else {
                return 2;
            }
        };
        CalendarBase.prototype.navigateTitle = function (e) {
            e.preventDefault();
            this.title(e);
            if (this.getModuleName() === 'calendar') {
                this.table.focus();
            }
        };
        CalendarBase.prototype.previous = function () {
            this.effect = '';
            var currentView = this.getViewNumber(this.currentView());
            switch (this.currentView()) {
                case 'Month':
                    this.addMonths(this.currentDate, -1);
                    this.switchView(currentView);
                    break;
                case 'Year':
                    this.addYears(this.currentDate, -1);
                    this.switchView(currentView);
                    break;
                case 'Decade':
                    this.addYears(this.currentDate, -10);
                    this.switchView(currentView);
                    break;
            }
        };
        CalendarBase.prototype.navigatePrevious = function (e) {
            e.preventDefault();
            this.previous();
            this.triggerNavigate(e);
            if (this.getModuleName() === 'calendar') {
                this.table.focus();
            }
        };
        CalendarBase.prototype.next = function () {
            this.effect = '';
            var currentView = this.getViewNumber(this.currentView());
            switch (this.currentView()) {
                case 'Month':
                    this.addMonths(this.currentDate, 1);
                    this.switchView(currentView);
                    break;
                case 'Year':
                    this.addYears(this.currentDate, 1);
                    this.switchView(currentView);
                    break;
                case 'Decade':
                    this.addYears(this.currentDate, 10);
                    this.switchView(currentView);
                    break;
            }
        };
        CalendarBase.prototype.navigateNext = function (eve) {
            eve.preventDefault();
            this.next();
            this.triggerNavigate(eve);
            if (this.getModuleName() === 'calendar') {
                this.table.focus();
            }
        };
        CalendarBase.prototype.navigateTo = function (view, date) {
            if (+date >= +this.min && +date <= +this.max) {
                this.currentDate = date;
            }
            if (+date <= +this.min) {
                this.currentDate = new Date('' + this.min);
            }
            if (+date >= +this.max) {
                this.currentDate = new Date('' + this.max);
            }
            if (this.getModuleName() === 'daterangepicker') {
                this.currentDate = date;
            }
            if ((this.getViewNumber(this.depth) >= this.getViewNumber(view))) {
                if ((this.getViewNumber(this.depth) <= this.getViewNumber(this.start))
                    || this.getViewNumber(this.depth) === this.getViewNumber(view)) {
                    view = this.depth;
                }
            }
            this.switchView(this.getViewNumber(view));
        };
        CalendarBase.prototype.currentView = function () {
            if (this.contentElement.classList.contains(YEAR)) {
                return 'Year';
            }
            else if (this.contentElement.classList.contains(DECADE)) {
                return 'Decade';
            }
            else {
                return 'Month';
            }
        };
        CalendarBase.prototype.getDateVal = function (date, value) {
            return (!ej2_base_7.isNullOrUndefined(value) && date.getDate() === (value).getDate()
                && date.getMonth() === (value).getMonth() && date.getFullYear() === (value).getFullYear());
        };
        CalendarBase.prototype.getCultureObjects = function (ld, c) {
            return ej2_base_6.getValue('main.' + '' + this.locale + '.dates.calendars.gregorian.days.format.short', ld);
        };
        ;
        CalendarBase.prototype.getWeek = function (d) {
            var currentDate = new Date('' + d).valueOf();
            var date = new Date(d.getFullYear(), 0, 1).valueOf();
            var a = (currentDate - date);
            return Math.ceil((((a) / dayMilliSeconds) + new Date(date).getDay() + 1) / 7);
        };
        CalendarBase.prototype.setStartDate = function (date, time) {
            var tzOffset = date.getTimezoneOffset();
            var d = new Date(date.getTime() + time);
            var tzOffsetDiff = d.getTimezoneOffset() - tzOffset;
            date.setTime(d.getTime() + tzOffsetDiff * minutesMilliSeconds);
        };
        CalendarBase.prototype.addMonths = function (date, i) {
            var day = date.getDate();
            date.setDate(1);
            date.setMonth(date.getMonth() + i);
            date.setDate(Math.min(day, this.getMaxDays(date)));
        };
        CalendarBase.prototype.addYears = function (date, i) {
            var day = date.getDate();
            date.setDate(1);
            date.setFullYear(date.getFullYear() + i);
            date.setDate(Math.min(day, this.getMaxDays(date)));
        };
        CalendarBase.prototype.getIdValue = function (e, element) {
            var eve;
            if (e) {
                eve = e.currentTarget;
            }
            else {
                eve = element;
            }
            var dateFormatOptions = { type: 'dateTime', skeleton: 'full' };
            var dateString = this.globalize.formatDate(new Date(parseInt('' + eve.getAttribute('id'), 0)), dateFormatOptions);
            var date = this.globalize.parseDate(dateString, dateFormatOptions);
            var value = date.valueOf() - date.valueOf() % 1000;
            return new Date(value);
        };
        CalendarBase.prototype.selectDate = function (e, date, node, value) {
            var element = node || e.currentTarget;
            if (this.currentView() === 'Decade') {
                this.setDateDecade(this.currentDate, date.getFullYear());
            }
            else if (this.currentView() === 'Year') {
                this.setDateYear(this.currentDate, date);
            }
            else {
                this.setProperties({ value: new Date('' + date) }, true);
                this.currentDate = new Date('' + date);
            }
            var tableBodyElement = ej2_base_5.closest(element, '.' + ROOT);
            if (ej2_base_7.isNullOrUndefined(tableBodyElement)) {
                tableBodyElement = this.tableBodyElement;
            }
            if (!ej2_base_7.isNullOrUndefined(tableBodyElement.querySelector('.' + SELECTED))) {
                ej2_base_5.removeClass([tableBodyElement.querySelector('.' + SELECTED)], SELECTED);
            }
            if (!ej2_base_7.isNullOrUndefined(tableBodyElement.querySelector('.' + FOCUSEDDATE))) {
                ej2_base_5.removeClass([tableBodyElement.querySelector('.' + FOCUSEDDATE)], FOCUSEDDATE);
            }
            ej2_base_5.addClass([element], SELECTED);
        };
        CalendarBase.prototype.setAriaActiveDescendant = function () {
            ej2_base_5.attributes(this.table, {
                'aria-activedescendant': '' + this.setActiveDescendant()
            });
        };
        CalendarBase.prototype.previousIconHandler = function (disabled) {
            if (disabled) {
                ej2_base_1.EventHandler.remove(this.previousIcon, 'click', this.navigatePreviousHandler);
                ej2_base_5.addClass([this.previousIcon], '' + DISABLED);
                ej2_base_5.addClass([this.previousIcon], '' + OVERLAY);
                this.previousIcon.setAttribute('aria-disabled', 'true');
            }
            else {
                ej2_base_1.EventHandler.add(this.previousIcon, 'click', this.navigatePreviousHandler);
                ej2_base_5.removeClass([this.previousIcon], '' + DISABLED);
                ej2_base_5.removeClass([this.previousIcon], '' + OVERLAY);
                this.previousIcon.setAttribute('aria-disabled', 'false');
            }
        };
        CalendarBase.prototype.renderDayCellEvent = function (args) {
            ej2_base_6.extend(this.renderDayCellArgs, { name: 'renderDayCell' });
            this.trigger('renderDayCell', args);
        };
        CalendarBase.prototype.navigatedEvent = function (eve) {
            ej2_base_6.extend(this.navigatedArgs, { name: 'navigated', event: eve });
            this.trigger('navigated', this.navigatedArgs);
        };
        CalendarBase.prototype.triggerNavigate = function (event) {
            this.navigatedArgs = { view: this.currentView(), date: this.currentDate };
            this.navigatedEvent(event);
        };
        CalendarBase.prototype.nextIconHandler = function (disabled) {
            if (disabled) {
                ej2_base_1.EventHandler.remove(this.nextIcon, 'click', this.navigateNextHandler);
                ej2_base_5.addClass([this.nextIcon], DISABLED);
                ej2_base_5.addClass([this.nextIcon], OVERLAY);
                this.nextIcon.setAttribute('aria-disabled', 'true');
            }
            else {
                ej2_base_1.EventHandler.add(this.nextIcon, 'click', this.navigateNextHandler);
                ej2_base_5.removeClass([this.nextIcon], DISABLED);
                ej2_base_5.removeClass([this.nextIcon], OVERLAY);
                this.nextIcon.setAttribute('aria-disabled', 'false');
            }
        };
        CalendarBase.prototype.compare = function (startDate, endDate, modifier) {
            var start = endDate.getFullYear();
            var end;
            var result;
            end = start;
            result = 0;
            if (modifier) {
                start = start - start % modifier;
                end = start - start % modifier + modifier - 1;
            }
            if (startDate.getFullYear() > end) {
                result = 1;
            }
            else if (startDate.getFullYear() < start) {
                result = -1;
            }
            return result;
        };
        CalendarBase.prototype.isMinMaxRange = function (date) {
            return +date >= +this.min && +date <= +this.max;
        };
        CalendarBase.prototype.compareYear = function (start, end) {
            return this.compare(start, end, 0);
        };
        CalendarBase.prototype.compareDecade = function (start, end) {
            return this.compare(start, end, 10);
        };
        CalendarBase.prototype.shiftArray = function (array, i) {
            return array.slice(i).concat(array.slice(0, i));
        };
        CalendarBase.prototype.addDay = function (date, i, e, max, min) {
            var column = i;
            var value = new Date(+date);
            if (!ej2_base_7.isNullOrUndefined(this.tableBodyElement) && !ej2_base_7.isNullOrUndefined(e)) {
                while (this.findNextTD(new Date(+date), column, max, min)) {
                    column += i;
                }
                var rangeValue = new Date(value.setDate(value.getDate() + column));
                column = (+rangeValue > +max || +rangeValue < +min) ? column === i ? i - i : i : column;
            }
            date.setDate(date.getDate() + column);
        };
        CalendarBase.prototype.findNextTD = function (date, column, max, min) {
            var value = new Date(date.setDate(date.getDate() + column));
            var collection = [];
            var isDisabled = false;
            if ((!ej2_base_7.isNullOrUndefined(value) && value.getMonth()) === (!ej2_base_7.isNullOrUndefined(this.currentDate) && this.currentDate.getMonth())) {
                var tdEles = this.renderDays(value, null);
                collection = tdEles.filter(function (element) {
                    return element.classList.contains(DISABLED);
                });
            }
            else {
                collection = this.tableBodyElement.querySelectorAll('td.' + DISABLED);
            }
            if (+value <= (+(max)) && +value >= (+(min))) {
                if (collection.length) {
                    for (var i = 0; i < collection.length; i++) {
                        isDisabled = (+value === +new Date(parseInt(collection[i].id, 0))) ? true : false;
                        if (isDisabled) {
                            break;
                        }
                    }
                }
            }
            return isDisabled;
        };
        CalendarBase.prototype.getMaxDays = function (d) {
            var date;
            var month;
            var tmpDate = new Date('' + d);
            date = 28;
            month = tmpDate.getMonth();
            while (tmpDate.getMonth() === month) {
                ++date;
                tmpDate.setDate(date);
            }
            return date - 1;
        };
        CalendarBase.prototype.setDateDecade = function (date, year) {
            date.setFullYear(year);
            this.setProperties({ value: new Date('' + date) }, true);
        };
        ;
        CalendarBase.prototype.setDateYear = function (date, value) {
            date.setFullYear(value.getFullYear(), value.getMonth(), date.getDate());
            if (value.getMonth() !== date.getMonth()) {
                date.setDate(0);
            }
            this.setProperties({ value: new Date('' + date) }, true);
            this.currentDate = new Date('' + value);
        };
        CalendarBase.prototype.compareMonth = function (start, end) {
            var result;
            if (start.getFullYear() > end.getFullYear()) {
                result = 1;
            }
            else if (start.getFullYear() < end.getFullYear()) {
                result = -1;
            }
            else {
                result = start.getMonth() === end.getMonth() ? 0 : start.getMonth() > end.getMonth() ? 1 : -1;
            }
            return result;
        };
        __decorate([
            ej2_base_7.Property(new Date(1900, 0, 1))
        ], CalendarBase.prototype, "min", void 0);
        __decorate([
            ej2_base_7.Property(new Date(2099, 11, 31))
        ], CalendarBase.prototype, "max", void 0);
        __decorate([
            ej2_base_7.Property(0)
        ], CalendarBase.prototype, "firstDayOfWeek", void 0);
        __decorate([
            ej2_base_7.Property('Month')
        ], CalendarBase.prototype, "start", void 0);
        __decorate([
            ej2_base_7.Property('Month')
        ], CalendarBase.prototype, "depth", void 0);
        __decorate([
            ej2_base_7.Property(false)
        ], CalendarBase.prototype, "weekNumber", void 0);
        __decorate([
            ej2_base_7.Property(true)
        ], CalendarBase.prototype, "showTodayButton", void 0);
        __decorate([
            ej2_base_7.Event()
        ], CalendarBase.prototype, "created", void 0);
        __decorate([
            ej2_base_7.Event()
        ], CalendarBase.prototype, "destroyed", void 0);
        __decorate([
            ej2_base_7.Event()
        ], CalendarBase.prototype, "navigated", void 0);
        __decorate([
            ej2_base_7.Event()
        ], CalendarBase.prototype, "renderDayCell", void 0);
        CalendarBase = __decorate([
            ej2_base_3.NotifyPropertyChanges
        ], CalendarBase);
        return CalendarBase;
    }(ej2_base_1.Component));
    exports.CalendarBase = CalendarBase;
    var Calendar = (function (_super) {
        __extends(Calendar, _super);
        function Calendar(options, element) {
            return _super.call(this, options, element) || this;
        }
        Calendar.prototype.render = function () {
            this.validateDate();
            this.minMaxUpdate();
            _super.prototype.render.call(this);
        };
        Calendar.prototype.validateDate = function () {
            if (typeof this.value === 'string') {
                this.setProperties({ value: this.checkDateValue(new Date('' + this.value)) }, true);
            }
            _super.prototype.validateDate.call(this, this.value);
            if (!ej2_base_7.isNullOrUndefined(this.value) && this.min <= this.max && this.value >= this.min && this.value <= this.max) {
                this.currentDate = new Date('' + this.value);
            }
            if (isNaN(+this.value)) {
                this.setProperties({ value: null }, true);
            }
        };
        Calendar.prototype.minMaxUpdate = function () {
            if (this.getModuleName() === 'calendar') {
                if (!ej2_base_7.isNullOrUndefined(this.value) && this.value <= this.min && this.min <= this.max) {
                    this.setProperties({ value: this.min }, true);
                    this.changedArgs = { value: this.value };
                }
                else {
                    if (!ej2_base_7.isNullOrUndefined(this.value) && this.value >= this.max && this.min <= this.max) {
                        this.setProperties({ value: this.max }, true);
                        this.changedArgs = { value: this.value };
                    }
                }
            }
            if (this.getModuleName() !== 'calendar') {
                if (!ej2_base_7.isNullOrUndefined(this.value) && this.value < this.min && this.min <= this.max) {
                    _super.prototype.minMaxUpdate.call(this, this.min);
                }
                else {
                    if (!ej2_base_7.isNullOrUndefined(this.value) && this.value > this.max && this.min <= this.max) {
                        _super.prototype.minMaxUpdate.call(this, this.max);
                    }
                }
            }
            else {
                _super.prototype.minMaxUpdate.call(this, this.value);
            }
        };
        Calendar.prototype.todayButtonClick = function () {
            if (this.showTodayButton) {
                var tempValue = new Date();
                if (this.value) {
                    tempValue.setHours(this.value.getHours());
                    tempValue.setMinutes(this.value.getMinutes());
                    tempValue.setSeconds(this.value.getSeconds());
                }
                else {
                    tempValue = new Date(tempValue.getFullYear(), tempValue.getMonth(), tempValue.getDate(), 0, 0, 0);
                }
                this.setProperties({ value: tempValue }, true);
                _super.prototype.todayButtonClick.call(this, new Date(+this.value));
            }
        };
        Calendar.prototype.keyActionHandle = function (e) {
            _super.prototype.keyActionHandle.call(this, e, this.value);
        };
        Calendar.prototype.preRender = function () {
            var _this = this;
            this.changeHandler = function (e) {
                _this.triggerChange(e);
            };
            _super.prototype.preRender.call(this, this.value);
        };
        ;
        Calendar.prototype.createContent = function () {
            this.previousDate = this.value;
            _super.prototype.createContent.call(this);
        };
        Calendar.prototype.minMaxDate = function (localDate) {
            return _super.prototype.minMaxDate.call(this, localDate);
        };
        Calendar.prototype.renderMonths = function (e) {
            _super.prototype.renderMonths.call(this, e, this.value);
        };
        Calendar.prototype.renderDays = function (currentDate, e) {
            return _super.prototype.renderDays.call(this, currentDate, e, this.value);
        };
        Calendar.prototype.renderYears = function (e) {
            _super.prototype.renderYears.call(this, e, this.value);
        };
        Calendar.prototype.renderDecades = function (e) {
            _super.prototype.renderDecades.call(this, e, this.value);
        };
        Calendar.prototype.renderTemplate = function (elements, count, classNm, e) {
            _super.prototype.renderTemplate.call(this, elements, count, classNm, e, this.value);
            this.changedArgs = { value: this.value };
            this.changeHandler();
        };
        Calendar.prototype.clickHandler = function (e) {
            var eve = e.currentTarget;
            if (eve.classList.contains(OTHERMONTH)) {
                this.value = this.getIdValue(e, null);
            }
            _super.prototype.clickHandler.call(this, e, this.value);
        };
        Calendar.prototype.switchView = function (view, e) {
            _super.prototype.switchView.call(this, view, e);
        };
        Calendar.prototype.getModuleName = function () {
            _super.prototype.getModuleName.call(this);
            return 'calendar';
        };
        Calendar.prototype.getPersistData = function () {
            _super.prototype.getPersistData.call(this);
            var keyEntity = ['value'];
            return this.addOnPersist(keyEntity);
        };
        Calendar.prototype.onPropertyChanged = function (newProp, oldProp) {
            this.effect = '';
            for (var _i = 0, _a = Object.keys(newProp); _i < _a.length; _i++) {
                var prop = _a[_i];
                switch (prop) {
                    case 'value':
                        if (typeof newProp.value === 'string') {
                            this.setProperties({ value: new Date('' + newProp.value) }, true);
                        }
                        else {
                            newProp.value = new Date('' + newProp.value);
                        }
                        if (isNaN(+this.value)) {
                            this.setProperties({ value: oldProp.value }, true);
                        }
                        this.validateDate();
                        this.minMaxUpdate();
                        _super.prototype.setValueUpdate.call(this);
                        break;
                    default:
                        _super.prototype.onPropertyChanged.call(this, newProp, oldProp);
                }
            }
        };
        Calendar.prototype.destroy = function () {
            _super.prototype.destroy.call(this);
        };
        Calendar.prototype.navigateTo = function (view, date) {
            this.minMaxUpdate();
            _super.prototype.navigateTo.call(this, view, date);
        };
        Calendar.prototype.currentView = function () {
            return _super.prototype.currentView.call(this);
        };
        Calendar.prototype.selectDate = function (e, date, element) {
            _super.prototype.selectDate.call(this, e, date, element, this.value);
            this.changedArgs = { value: this.value };
            this.changeHandler(e);
        };
        Calendar.prototype.changeEvent = function (e) {
            this.trigger('change', this.changedArgs);
            this.previousDate = new Date(+this.value);
        };
        Calendar.prototype.triggerChange = function (e) {
            this.changedArgs.event = this.changedArgs.e = e || null;
            this.changedArgs.isInteracted = !ej2_base_7.isNullOrUndefined(e);
            if (!ej2_base_7.isNullOrUndefined(this.value)) {
                this.setProperties({ value: this.value }, true);
            }
            if (+this.value !== Number.NaN && +this.value !== +this.previousDate) {
                this.changeEvent(e);
            }
        };
        __decorate([
            ej2_base_7.Property(null)
        ], Calendar.prototype, "value", void 0);
        __decorate([
            ej2_base_7.Event()
        ], Calendar.prototype, "change", void 0);
        Calendar = __decorate([
            ej2_base_3.NotifyPropertyChanges
        ], Calendar);
        return Calendar;
    }(CalendarBase));
    exports.Calendar = Calendar;
});
