// @syncfusion/ej2-grids/src/grid/base/constant.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.created = 'create';
    exports.destroyed = 'destroy';
    exports.load = 'load';
    exports.rowDataBound = 'rowDataBound';
    exports.queryCellInfo = 'queryCellInfo';
    exports.actionBegin = 'actionBegin';
    exports.actionComplete = 'actionComplete';
    exports.actionFailure = 'actionFailure';
    exports.dataBound = 'dataBound';
    exports.rowSelecting = 'rowSelecting';
    exports.rowSelected = 'rowSelected';
    exports.rowDeselecting = 'rowDeselecting';
    exports.rowDeselected = 'rowDeselected';
    exports.cellSelecting = 'cellSelecting';
    exports.cellSelected = 'cellSelected';
    exports.cellDeselecting = 'cellDeselecting';
    exports.cellDeselected = 'cellDeselected';
    exports.columnDragStart = 'columnDragStart';
    exports.columnDrag = 'columnDrag';
    exports.columnDrop = 'columnDrop';
    exports.rowDragStart = 'rowDragStart';
    exports.rowDrag = 'rowDrag';
    exports.rowDrop = 'rowDrop';
    exports.beforePrint = 'beforePrint';
    exports.printComplete = 'printComplete';
    exports.detailDataBound = 'detailDataBound';
    exports.toolbarClick = 'toolbarClick';
    exports.batchAdd = 'batchAdd';
    exports.batchCancel = 'batchCancel';
    exports.batchDelete = 'batchDelete';
    exports.beforeBatchAdd = 'beforeBatchAdd';
    exports.beforeBatchDelete = 'beforeBatchDelete';
    exports.beforeBatchSave = 'beforeBatchSave';
    exports.beginEdit = 'beginEdit';
    exports.cellEdit = 'cellEdit';
    exports.cellSave = 'cellSave';
    exports.endAdd = 'endAdd';
    exports.endDelete = 'endDelete';
    exports.endEdit = 'endEdit';
    exports.recordDoubleClick = 'recordDoubleClick';
    exports.recordClick = 'recordClick';
    exports.beforeDataBound = 'beforeDataBound';
    exports.beforeOpenColumnChooser = 'beforeOpenColumnChooser';
    exports.resizeStart = 'resizeStart';
    exports.onResize = 'resizing';
    exports.resizeStop = 'resizeStop';
    exports.checkBoxChange = 'checkBoxChange';
    exports.beforeCopy = 'beforeCopy';
    exports.filterChoiceRequest = 'filterchoicerequest';
    exports.filterAfterOpen = 'filterafteropen';
    exports.filterBeforeOpen = 'filterbeforeopen';
    exports.filterSearchBegin = 'filtersearchbegin';
    exports.initialLoad = 'initial-load';
    exports.initialEnd = 'initial-end';
    exports.dataReady = 'data-ready';
    exports.contentReady = 'content-ready';
    exports.uiUpdate = 'ui-update';
    exports.onEmpty = 'on-empty';
    exports.inBoundModelChanged = 'inbound-model-changed';
    exports.modelChanged = 'model-changed';
    exports.colGroupRefresh = 'colgroup-refresh';
    exports.headerRefreshed = 'header-refreshed';
    exports.pageBegin = 'paging-begin';
    exports.pageComplete = 'paging-complete';
    exports.sortBegin = 'sorting-begin';
    exports.sortComplete = 'sorting-complete';
    exports.filterBegin = 'filtering-begin';
    exports.filterComplete = 'filtering-complete';
    exports.searchBegin = 'searching-begin';
    exports.searchComplete = 'searching-complete';
    exports.reorderBegin = 'reorder-begin';
    exports.reorderComplete = 'reorder-complete';
    exports.rowDragAndDropBegin = 'rowdraganddrop-begin';
    exports.rowDragAndDropComplete = 'rowdraganddrop-complete';
    exports.groupBegin = 'grouping-begin';
    exports.groupComplete = 'grouping-complete';
    exports.ungroupBegin = 'ungrouping-begin';
    exports.ungroupComplete = 'ungrouping-complete';
    exports.rowSelectionBegin = 'rowselecting';
    exports.rowSelectionComplete = 'rowselected';
    exports.columnSelectionBegin = 'columnselecting';
    exports.columnSelectionComplete = 'columnselected';
    exports.cellSelectionBegin = 'cellselecting';
    exports.cellSelectionComplete = 'cellselected';
    exports.beforeCellFocused = 'beforecellfocused';
    exports.cellFocused = 'cellfocused';
    exports.keyPressed = 'key-pressed';
    exports.click = 'click';
    exports.destroy = 'destroy';
    exports.columnVisibilityChanged = 'column-visible-changed';
    exports.scroll = 'scroll';
    exports.columnWidthChanged = 'column-width-changed';
    exports.columnPositionChanged = 'column-position-changed';
    exports.rowDragAndDrop = 'row-drag-and-drop';
    exports.rowsAdded = 'rows-added';
    exports.rowsRemoved = 'rows-removed';
    exports.columnDragStop = 'column-drag-stop';
    exports.headerDrop = 'header-drop';
    exports.dataSourceModified = 'datasource-modified';
    exports.refreshComplete = 'refresh-complete';
    exports.refreshVirtualBlock = 'refresh-virtual-block';
    exports.dblclick = 'dblclick';
    exports.toolbarRefresh = 'toolbar-refresh';
    exports.bulkSave = 'bulk-save';
    exports.autoCol = 'auto-col';
    exports.tooltipDestroy = 'tooltip-destroy';
    exports.updateData = 'update-data';
    exports.editBegin = 'edit-begin';
    exports.editComplete = 'edit-complete';
    exports.addBegin = 'add-begin';
    exports.addComplete = 'add-complete';
    exports.saveComplete = 'save-complete';
    exports.deleteBegin = 'delete-begin';
    exports.deleteComplete = 'delete-complete';
    exports.preventBatch = 'prevent-batch';
    exports.dialogDestroy = 'dialog-destroy';
    exports.crudAction = 'crud-Action';
    exports.addDeleteAction = 'add-delete-Action';
    exports.destroyForm = 'destroy-form';
    exports.doubleTap = 'double-tap';
    exports.beforeExcelExport = 'beforeExcelExport';
    exports.excelExportComplete = 'excelExportComplete';
    exports.excelQueryCellInfo = 'excelQueryCellInfo';
    exports.beforePdfExport = 'beforePdfExport';
    exports.pdfExportComplete = 'pdfExportComplete';
    exports.pdfQueryCellInfo = 'pdfQueryCellInfo';
    exports.accessPredicate = 'access-predicate';
    exports.contextMenuClick = 'contextMenuClick';
    exports.freezeRender = 'freezerender';
    exports.freezeRefresh = 'freezerefresh';
    exports.contextMenuOpen = 'contextMenuOpen';
    exports.columnMenuClick = 'columnMenuClick';
    exports.columnMenuOpen = 'columnMenuOpen';
    exports.filterOpen = 'filterOpen';
    exports.filterDialogCreated = 'filterDialogCreated';
    exports.filterMenuClose = 'filter-menu-close';
    exports.initForeignKeyColumn = 'initForeignKeyColumn';
    exports.getForeignKeyData = 'getForeignKeyData';
    exports.generateQuery = 'generateQuery';
    exports.showEmptyGrid = 'showEmptyGrid';
    exports.foreignKeyData = 'foreignKeyData';
    exports.dataStateChange = 'dataStateChange';
    exports.dataSourceChanged = 'dataSourceChanged';
    exports.rtlUpdated = 'rtl-updated';
    exports.beforeFragAppend = 'beforeFragAppend';
});
